{
 "metadata": {
  "name": "",
  "signature": "sha256:d0c019b9f6307a62500770173bac98dc8f1cbad2057b389264fe1437e561fd59"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Libraries"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pandas as pd\n",
      "import os, sys, gzip"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import imp\n",
      "\n",
      "pandasVCF = imp.load_source('module.name', '/path/to/file.py')\n",
      "foo.MyClass()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 55
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Data"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_path = '../test_data/NA128_17member_pedigree_10k.vcf.gz'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 64
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Functions"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "class Vcf_metadata(object):\n",
      "    '''\n",
      "    This class parses a VCF header into a pandas\n",
      "    dataframe object.  It recognizes gzip and uncompressed\n",
      "    file formats.  This function assumes the header does not\n",
      "    extent past 5000 lines\n",
      "    \n",
      "    '''\n",
      "    def __init__(self, filename):\n",
      "        \n",
      "        if filename.endswith('.gz'):\n",
      "            self.compression = 'gzip'\n",
      "            if filename+'.tbi' in os.listdir(os.path.split(filename)[0]):\n",
      "                header_lines = os.popen('tabix -H ' + filename).readlines()\n",
      "                self.header = [l.replace('#CHROM','CHROM') for l in header_lines if l.startswith('#')]\n",
      "            os.system('tabix -p vcf ' + filename)\n",
      "            header_lines = os.popen('tabix -H ' + filename).readlines()\n",
      "            self.header = [l for l in header_lines if l.startswith('#')]\n",
      "            return 0\n",
      "        \n",
      "        else:\n",
      "            self.compression = ''\n",
      "            header_lines = os.popen('head -5000 ' + filename).readlines()\n",
      "            self.header = [l for l in header_lines if l.startswith('#')]\n",
      "            return 0\n",
      "            \n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_multiallelic_bases(df, sample_col, single_sample_vcf=True):\n",
      "    '''\n",
      "    This function parses multiallele variants into DNA base representations.\n",
      "    It currently does not support haploid chromosomes.\n",
      "    '''\n",
      "    \n",
      "    def _get_allele(line, gt_col):\n",
      "        '''\n",
      "        Returns allele base call from multi-allelic variants\n",
      "        '''\n",
      "        \n",
      "        alleles = [line['REF']]\n",
      "        alleles.extend(line['ALT'].split(\",\"))\n",
      "        a1 = \".\"\n",
      "        try:\n",
      "            a1 = alleles[int(line[gt_col])]  #returns missing if gt_int_call is \".\"\n",
      "        except:\n",
      "            a1 = \".\"\n",
      "        return a1\n",
      "\n",
      "\n",
      "    def get_GT_multisample_vcf(line, sample_col, gt_index):\n",
      "        return int( line[sample_col].split(line['phase'])[int(gt_index)])\n",
      "    \n",
      "    \n",
      "    if single_sample_vcf:\n",
      "        df['phase'] = df[sample_col].str[1]\n",
      "        df = df[df['phase']!=':']  #removing haploid variants\n",
      "        \n",
      "        df['GT1'] = df[sample_col].str[0]\n",
      "        df = df[df['GT1']!='.']  #removing variants with missing calls\n",
      "        df['GT1'] = df['GT1'].astype(int)\n",
      "        \n",
      "        df['GT2'] = df[sample_col].str[2]\n",
      "        df = df[df['GT2']!='.']  #removing variants with missing calls\n",
      "        df['GT2'] = df['GT2'].astype(int)\n",
      "\n",
      "\n",
      "    if not single_sample_vcf:\n",
      "        df['phase'] = df.apply(get_phase, args=['GT'], axis=1)  #get phase\n",
      "        df = df[df.phase != \"-\"]  #likley occurs at sex chromosome sites\n",
      "        df['GT1'] = df.apply(get_GT_multisample, args=[sample_col, 0], axis=1)\n",
      "        df['GT2'] = df.apply(get_GT_multisample, args=[sample_col, 1], axis=1)\n",
      "\n",
      "    \n",
      "    \n",
      "    #SLOW PROCESS MULTIPLE ALLELE GENOTYPES\n",
      "    df_multi = df[(df.GT1.astype(int)>1) | (df.GT2.astype(int)>1)] #select all multi-alleleic variants\n",
      "    df['a1'] = df.apply(_get_allele, args=['GT1'], axis=1)\n",
      "    df['a2'] = df.apply(_get_allele, args=['GT2'], axis=1)\n",
      "\n",
      "    return df\n",
      "\n",
      "\n",
      "\n",
      "def get_biallelic_bases(df, sample_col, single_sample_vcf=True):\n",
      "    '''\n",
      "    This function is 10X faster than previous iterations\n",
      "\n",
      "    '''\n",
      "\n",
      "    def _get_allele(line, gt_col):\n",
      "        '''\n",
      "        Returns allelic base, handles multi-allelic variants\n",
      "        '''\n",
      "        alleles = [line['REF']]\n",
      "        alleles.extend(line['ALT'].split(\",\"))\n",
      "        a1 = \".\"\n",
      "        try:\n",
      "            a1 = alleles[int(line[gt_col])]  #returns missing if gt_int_call is \".\"\n",
      "        except:\n",
      "            a1 = \".\"\n",
      "        return a1\n",
      "\n",
      "\n",
      "    def get_GT_multisample_vcf(line, sample_col, gt_index):\n",
      "        return int( line[sample_col].split(line['phase'])[int(gt_index)])\n",
      "\n",
      "\n",
      "    if single_sample_vcf:\n",
      "        df['phase'] = df[sample_col].str[1]\n",
      "        df = df[df['phase']!=':']\n",
      "        df['GT1'] = df[sample_col].str[0]\n",
      "        df = df[df['GT1']!='.']\n",
      "        df['GT1'] = df['GT1'].astype(int)\n",
      "        df['GT2'] = df[sample_col].str[2]\n",
      "        df = df[df['GT2']!='.']\n",
      "        df['GT2'] = df['GT2'].astype(int)\n",
      "\n",
      "\n",
      "    if not single_sample_vcf:\n",
      "        df['phase'] = df.apply(get_phase, args=['GT'], axis=1)  #get phase\n",
      "        df = df[df.phase != \"-\"]  #likley occurs at sex chromosome sites\n",
      "        df['GT1'] = df.apply(get_GT_multisample, args=[sample_col, 0], axis=1)\n",
      "        df['GT2'] = df.apply(get_GT_multisample, args=[sample_col, 1], axis=1)\n",
      "\n",
      "\n",
      "    #FAST PROCESS SIMPLE ALLELE GENOTYPES\n",
      "    df_simple = df\n",
      "    \n",
      "    df_gt1_ref = df_simple[df_simple.GT1==0][['REF']]  #get a1 ref alleles\n",
      "    df_gt1_ref.columns = ['a1']\n",
      "    df_gt2_ref = df_simple[df_simple.GT2==0][['REF']]  #get a2 ref alleles\n",
      "    df_gt2_ref.columns = ['a2']\n",
      "\n",
      "\n",
      "    df_gt1_alt = df_simple[df_simple.GT1==1][['ALT']]  #get a1 alt alleles\n",
      "    df_gt1_alt.columns = ['a1']\n",
      "    df_gt2_alt = df_simple[df_simple.GT2==1][['ALT']]  #get a2 alt alleles\n",
      "    df_gt2_alt.columns = ['a2']\n",
      "\n",
      "\n",
      "    gt1_alleles = pd.concat([df_gt1_ref,df_gt1_alt])  #merging GT1 allele bases into a single df\n",
      "    #del gt1_alleles[0]\n",
      "    gt2_alleles = pd.concat([df_gt2_ref,df_gt2_alt])  #merging GT2 allele bases into a single df\n",
      "    #del gt2_alleles[0]\n",
      "    gt1_2_allele_df = gt1_alleles.join(gt2_alleles, how='outer')  #Joining the GT1 and GT2 simple allele bases \n",
      "\n",
      "\n",
      "    df = df.join(gt1_2_allele_df, how='inner')  #Adding simle allele a1 and a2 columns to original df\n",
      "\n",
      "    return df\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "def zygosity_fast(df):\n",
      "    '''\n",
      "    This function quickly assigns zygosity states to \n",
      "    each variant using set logic\n",
      "    '''\n",
      "\n",
      "    df_hom_ref = df[ (df['a1'] == df['REF']) & (df['a2'] == df['REF'])]\n",
      "    df_hom_ref['zygosity'] = 'hom-ref'\n",
      "\n",
      "    df_hom_miss = df[ (df['a1'] == '.') & (df['a2'] == '.')]\n",
      "    df_hom_miss['zygosity'] = 'hom-miss'\n",
      "\n",
      "    df_het_miss = df[ (df['a1'] == '.') | (df['a2'] == '.') ]\n",
      "    df_het_miss['zygosity'] = 'het-miss'\n",
      "\n",
      "\n",
      "    df_not_miss = df[~df.index.isin( set(df_hom_miss.index) | set(df_het_miss.index))  ]\n",
      "\n",
      "\n",
      "    df_het_alt = df_not_miss[ ((df_not_miss['a1'] != df_not_miss['REF']) & (df_not_miss['a2'] != df_not_miss['REF'])) & (df_not_miss['a1'] != df_not_miss['a2']) ]\n",
      "    df_het_alt['zygosity'] = 'het-alt'\n",
      "\n",
      "    df_hom_alt = df_not_miss[ (((df_not_miss['a1'] != df_not_miss['REF']) & (df_not_miss['a2'] != df_not_miss['REF']))) & (df_not_miss['a1'] == df_not_miss['a2']) ]\n",
      "    df_hom_alt['zygosity'] = 'hom-alt'\n",
      "\n",
      "    df_het_ref = df_not_miss[ ((df_not_miss['a1'] == df_not_miss['REF']) & (df_not_miss['a2'] != df_not_miss['REF'])) | ((df_not_miss['a1'] != df_not_miss['REF']) & (df_not_miss['a2'] == df_not_miss['REF']))  ]\n",
      "    df_het_ref['zygosity'] = 'het-ref'\n",
      "\n",
      "\n",
      "\n",
      "    df_zygosity = pd.concat([df_hom_ref, df_hom_miss, df_het_miss, df_het_ref, df_het_alt, df_hom_alt])\n",
      "    #return df_zygosity\n",
      "    assert len(df_zygosity) == len(df)\n",
      "    return df_zygosity\n",
      "\n",
      "\n",
      "def vartype_map(ref_alt_bases):\n",
      "    '''\n",
      "    This function assigns the following vartypes to the \n",
      "    allele specified by allele_base_col: snp, mnp, ins, del, indel or SV\n",
      "\n",
      "    '''\n",
      "    ref, alt = ref_alt_bases\n",
      "    len_diff = len(ref) - len(alt)\n",
      "\n",
      "    if ref == alt: return 'ref' #Orderd by frequency of the variant to reduce complexity\n",
      "\n",
      "    if len_diff == 0:\n",
      "        base_diff = [nt for i,nt in enumerate(alt) if ref[i] != alt[i]]\n",
      "        if len(base_diff) == 1: return 'snp' \n",
      "        else: return 'mnp'\n",
      "\n",
      "    if len_diff > 0:\n",
      "        base_diff = [nt for i,nt in enumerate(alt) if ref[i] != alt[i]]\n",
      "        if len(base_diff) > 0: return 'indel'\n",
      "        else: return 'del'\n",
      "\n",
      "    if len_diff < 0:\n",
      "        #base_diff = [nt for i,nt in enumerate(ref) if ref[i] != alt[i]]\n",
      "        return 'ins'\n",
      "\n",
      "    elif is_sv(ref,alt): return 'sv'\n",
      "\n",
      "    else: return 'indel or SV'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "class Vcf(object):\n",
      "    '''\n",
      "    Loads in a vcf file, aware of gzipped files.\n",
      "    '''\n",
      "    \n",
      "    def __init__(self, filename, sample_id='', cols='', chunksize=5000):\n",
      "        \n",
      "        #Header\n",
      "        header_parsed = Vcf_metadata(filename)\n",
      "        self.header_df = self.get_header_df(header_parsed.header)  #header parsed into key/values dataframe\n",
      "    \n",
      "\n",
      "        \n",
      "        #Sample IDs\n",
      "        self.samples = list(self.header_df.ix['SampleIDs'])[0]\n",
      "        self.sample_id = sample_id\n",
      "        \n",
      "        \n",
      "        #Columns\n",
      "        self.all_columns = list(self.header_df.ix['ColumnHeader'])[0]\n",
      "        self.FORMAT = self.all_columns[8]\n",
      "        \n",
      "        \n",
      "        assert len(set(cols) & set(['#CHROM', 'POS', 'REF', 'ALT', 'FORMAT'])), \"cols requires the following columns: ['#CHROM', 'POS', 'REF', 'ALT', 'FORMAT']\"    \n",
      "        self.cols = cols\n",
      "        if len(cols) > 0:\n",
      "            self.usecols = ['#CHROM']\n",
      "            self.usecols = [c for c in self.all_columns if c in cols]\n",
      "            if len(sample_id) > 0:\n",
      "                self.usecols.extend(sample_id)\n",
      "        else:\n",
      "            self.usecols = [s for s in self.cols if s not in self.samples]\n",
      "            self.usecols.extend(sample_id)\n",
      "        \n",
      "        #Open pandas chunk object (TextReader)\n",
      "        self.chunksize = chunksize\n",
      "        self.vcf_chunks = pd.read_table(filename, sep=\"\\t\", compression=header_parsed.compression, skiprows=(len(self.header_df)-2), usecols=self.usecols, chunksize=chunksize)\n",
      "        \n",
      "        \n",
      "        \n",
      "    def get_header_df(self, header_txt):\n",
      "        '''\n",
      "        Parses header into pandas DataFrame\n",
      "        '''\n",
      "        key_value_header = [i.replace('##','').replace('\\n','').split('=',1) for i in header_txt if '##' in i]\n",
      "        key_value_header.append(['SampleIDs',header_txt[-1].rstrip('\\n').split('\\t')[9:]])\n",
      "        key_value_header.append(['ColumnHeader', header_txt[-1].rstrip('\\n').split('\\t')])\n",
      "        header_df =  pd.DataFrame.from_records(key_value_header)\n",
      "        header_df.set_index(0,inplace=True)\n",
      "        header_df.index.name = 'header_keys'\n",
      "        header_df.columns = ['header_values']\n",
      "        return header_df\n",
      "    \n",
      "\n",
      "    \n",
      "    \n",
      "    def get_vcf_df_chunk(self):\n",
      "        '''\n",
      "        This function iterates through the VCF files using the user-defined \n",
      "        chunksize (default = 5000 lines).\n",
      "        '''\n",
      "        \n",
      "        self.df = self.vcf_chunks.get_chunk()\n",
      "        self.df.columns = [c.replace('#', '') for c in self.usecols]\n",
      "        self.df.set_index(['CHROM', 'POS', 'REF', 'ALT'], inplace=True, drop=False)\n",
      "        \n",
      "        self.df_bytes = self.df.values.nbytes + self.df.index.nbytes + self.df.columns.nbytes\n",
      "        \n",
      "        return 0\n",
      "\n",
      "\n",
      "    \n",
      "    \n",
      "    def parse_single_genotype_data(self, sample_id, split_cols=''):\n",
      "        '''\n",
      "        This function parses the genotype sample column and left joins to\n",
      "        the df object.\n",
      "\n",
      "        split_cols is a dictionary specifying the name of the column and the number\n",
      "        of values expected for that column, e.g. {'AD':2, 'PL':3}\n",
      "        '''\n",
      "        \n",
      "        df = self.df\n",
      "        genotypes = df.groupby(by='FORMAT')  #genotypes grouped by FORMAT variant annotations\n",
      "\n",
      "        #Iterate through genotype groups, dropping missing calls\n",
      "        master_df = []\n",
      "        for name,group in genotypes:\n",
      "            temp_group = group[sample_id]  #group of interest\n",
      "            #del temp_group['FORMAT']  #remove the format column\n",
      "            temp_group.replace(to_replace='.', value=nan, inplace=True)  #replace . with none, allows stack to remove null columns, space savings\n",
      "\n",
      "            temp_group_data = pd.DataFrame.from_records(list(temp_group.str.split(':')))\n",
      "            temp_group_data.index = temp_group.index\n",
      "            temp_group_data.columns = name.split(':')\n",
      "            temp_group_data.replace(to_replace='.', value=nan, inplace=True)\n",
      "\n",
      "            master_df.append(temp_group_data)\n",
      "\n",
      "        #Concatenating all genotype groups\n",
      "        sample_df = pd.concat(master_df)\n",
      "        sample_df.index.names = ['CHROM', 'POS', 'REF', 'ALT']\n",
      "\n",
      "        #spliting user-defined columns\n",
      "        if split_cols != '':\n",
      "            for col in split_cols:\n",
      "                for i in range(0, split_cols[col]):\n",
      "                    sample_df[col + '_' + str(i)] = sample_df[col].str.split(',').str[i]\n",
      "                del sample_df[col]\n",
      "        return sample_df\n",
      "    \n",
      "    \n",
      "    \n",
      "    \n",
      "    def get_vcf_annotations(self, sample_name, split_columns=''):\n",
      "        '''\n",
      "        This function adds the following annotations for each variant:\n",
      "        multiallele, phase, a1, a2, GT1, GT2, vartype1, vartype2, zygosity,\n",
      "        and parsed FORMAT values, see below for additional information.\n",
      "                        \n",
      "        Parameters\n",
      "        --------------\n",
      "        sample_name: str, required \n",
      "                    sample column header id, e.g. NA12878\n",
      "        \n",
      "        split_columns: dict, optional\n",
      "                    key:FORMAT id value:#fields expected\n",
      "                    e.g. {'AD':2} indicates Allelic Depth should be\n",
      "                    split into 2 columns.\n",
      "        \n",
      "        Output\n",
      "        --------------\n",
      "        This function adds the following annotations to each variant:\n",
      "        \n",
      "        multiallele: {0,1} 0=biallele  1=multiallelic\n",
      "    \n",
      "        phase: {'/', '|'} /=unphased, |=phased\n",
      "        \n",
      "        a1: DNA base representation of allele1 call, e.g. A\n",
      "        a2: DNA base representation of allele2 call, e.g. A\n",
      "        \n",
      "        GT1: numeric representation of allele1 call, e.g. 0\n",
      "        GT2: numeric representation of allele2 call, e.g. 1\n",
      "        \n",
      "        vartype1: {snp, mnp, ins, del, indel or SV} variant type of first allele\n",
      "        vartype2: {snp, mnp, ins, del, indel or SV} variant type of second allele\n",
      "        \n",
      "        zygosity: {het-ref, hom-ref, alt-ref, het-miss, hom-miss}\n",
      "        \n",
      "        FORMAT values: any values associated with the genotype calls are \n",
      "                        added as additional columns, split_columns are further\n",
      "                        split by ',' into individual columns\n",
      "        \n",
      "        \n",
      "        \n",
      "        '''\n",
      "        \n",
      "        self.df['multiallele'] = self.df.ALT.map(lambda x: 1 if \",\" in x else 0)  #1 is more than 1 alt allele, 0 else\n",
      "        self.multidf = self.df[self.df['multiallele'] == 1]\n",
      "        self.multidf = get_multiallelic_bases(self.multidf, sample_name)\n",
      "        \n",
      "        self.df = self.df[~self.df.index.isin(self.multidf.index)]\n",
      "        self.df = get_biallelic_bases(self.df, sample_name)\n",
      "        self.sample_name = sample_name\n",
      "        \n",
      "        \n",
      "        self.df = self.df.append(self.multidf)\n",
      "        \n",
      "        self.df = zygosity_fast(self.df)\n",
      "        self.df['vartype1'] = map(vartype_map, self.df[['REF','a1']].values)\n",
      "        self.df['vartype2'] = map(vartype_map, self.df[['REF','a2']].values)\n",
      "        \n",
      "        self.df.set_index(['CHROM', 'POS', 'REF', 'ALT'], inplace=True)\n",
      "        \n",
      "        self.df.sortlevel(level=['CHROM','POS','REF','ALT'],inplace=True)  #sorting biallelic and multiallele variants \n",
      "        \n",
      "        self.split_columns = split_columns\n",
      "        self.df = self.df.join( self.parse_single_genotype_data(self.sample_name, split_cols=split_columns), how='left' )\n",
      "        del self.df[sample_name]\n",
      "        \n",
      "        return self.df\n",
      "    \n",
      "    \n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 58
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#vcf_path = '/Users/ers_vader/datasets_raw/1000genomes/phase3/ALL.chr15.phase3_shapeit2_mvncall_integrated_v4.20130502.genotypes.vcf.gz'\n",
      "test = Vcf(vcf_path, cols=['#CHROM', 'POS', 'REF', 'ALT','FORMAT', 'NA12878'], chunksize=200)\n",
      "\n",
      "#simple = test.get_vcf_annotations('HG00096', split_cols={'AD':2, 'PL':3})"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 70
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "test.get_vcf_df_chunk()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 78,
       "text": [
        "0"
       ]
      }
     ],
     "prompt_number": 78
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "test.df.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th>CHROM</th>\n",
        "      <th>POS</th>\n",
        "      <th>REF</th>\n",
        "      <th>ALT</th>\n",
        "      <th>FORMAT</th>\n",
        "      <th>NA12878</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CHROM</th>\n",
        "      <th>POS</th>\n",
        "      <th>REF</th>\n",
        "      <th>ALT</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th rowspan=\"5\" valign=\"top\">chr1</th>\n",
        "      <th>57376</th>\n",
        "      <th>C</th>\n",
        "      <th>T</th>\n",
        "      <td> chr1</td>\n",
        "      <td> 57376</td>\n",
        "      <td> C</td>\n",
        "      <td> T</td>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td>                                     .</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>57856</th>\n",
        "      <th>T</th>\n",
        "      <th>A</th>\n",
        "      <td> chr1</td>\n",
        "      <td> 57856</td>\n",
        "      <td> T</td>\n",
        "      <td> A</td>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td> 0/1:22:99:49:0.347:249,0,524:99:32,17</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>57952</th>\n",
        "      <th>A</th>\n",
        "      <th>C</th>\n",
        "      <td> chr1</td>\n",
        "      <td> 57952</td>\n",
        "      <td> A</td>\n",
        "      <td> C</td>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td>    1/1:12:5.99:14:1.000:49,6,0:6:0,14</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>57999</th>\n",
        "      <th>G</th>\n",
        "      <th>T</th>\n",
        "      <td> chr1</td>\n",
        "      <td> 57999</td>\n",
        "      <td> G</td>\n",
        "      <td> T</td>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td>                                     .</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>58176</th>\n",
        "      <th>G</th>\n",
        "      <th>A</th>\n",
        "      <td> chr1</td>\n",
        "      <td> 58176</td>\n",
        "      <td> G</td>\n",
        "      <td> A</td>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td>                                     .</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 79,
       "text": [
        "                    CHROM    POS REF ALT                    FORMAT  \\\n",
        "CHROM POS   REF ALT                                                  \n",
        "chr1  57376 C   T    chr1  57376   C   T  GT:MQ:GQ:DP:VF:PL:GQX:AD   \n",
        "      57856 T   A    chr1  57856   T   A  GT:MQ:GQ:DP:VF:PL:GQX:AD   \n",
        "      57952 A   C    chr1  57952   A   C  GT:MQ:GQ:DP:VF:PL:GQX:AD   \n",
        "      57999 G   T    chr1  57999   G   T  GT:MQ:GQ:DP:VF:PL:GQX:AD   \n",
        "      58176 G   A    chr1  58176   G   A  GT:MQ:GQ:DP:VF:PL:GQX:AD   \n",
        "\n",
        "                                                   NA12878  \n",
        "CHROM POS   REF ALT                                         \n",
        "chr1  57376 C   T                                        .  \n",
        "      57856 T   A    0/1:22:99:49:0.347:249,0,524:99:32,17  \n",
        "      57952 A   C       1/1:12:5.99:14:1.000:49,6,0:6:0,14  \n",
        "      57999 G   T                                        .  \n",
        "      58176 G   A                                        .  "
       ]
      }
     ],
     "prompt_number": 79
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "testing = test.get_vcf_annotations('NA12878')\n",
      "testing.head(5)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th>FORMAT</th>\n",
        "      <th>multiallele</th>\n",
        "      <th>phase</th>\n",
        "      <th>GT1</th>\n",
        "      <th>GT2</th>\n",
        "      <th>a1</th>\n",
        "      <th>a2</th>\n",
        "      <th>zygosity</th>\n",
        "      <th>vartype1</th>\n",
        "      <th>vartype2</th>\n",
        "      <th>GT</th>\n",
        "      <th>MQ</th>\n",
        "      <th>GQ</th>\n",
        "      <th>DP</th>\n",
        "      <th>VF</th>\n",
        "      <th>PL</th>\n",
        "      <th>GQX</th>\n",
        "      <th>AD</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CHROM</th>\n",
        "      <th>POS</th>\n",
        "      <th>REF</th>\n",
        "      <th>ALT</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th rowspan=\"5\" valign=\"top\">chr1</th>\n",
        "      <th>57856</th>\n",
        "      <th>T</th>\n",
        "      <th>A</th>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1</td>\n",
        "      <td> T</td>\n",
        "      <td> A</td>\n",
        "      <td> het-ref</td>\n",
        "      <td> ref</td>\n",
        "      <td> snp</td>\n",
        "      <td> 0/1</td>\n",
        "      <td> 22</td>\n",
        "      <td>    99</td>\n",
        "      <td> 49</td>\n",
        "      <td> 0.347</td>\n",
        "      <td> 249,0,524</td>\n",
        "      <td> 99</td>\n",
        "      <td> 32,17</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>57952</th>\n",
        "      <th>A</th>\n",
        "      <th>C</th>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 1</td>\n",
        "      <td> 1</td>\n",
        "      <td> C</td>\n",
        "      <td> C</td>\n",
        "      <td> hom-alt</td>\n",
        "      <td> snp</td>\n",
        "      <td> snp</td>\n",
        "      <td> 1/1</td>\n",
        "      <td> 12</td>\n",
        "      <td>  5.99</td>\n",
        "      <td> 14</td>\n",
        "      <td> 1.000</td>\n",
        "      <td>    49,6,0</td>\n",
        "      <td>  6</td>\n",
        "      <td>  0,14</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>58211</th>\n",
        "      <th>A</th>\n",
        "      <th>G</th>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 1</td>\n",
        "      <td> 1</td>\n",
        "      <td> G</td>\n",
        "      <td> G</td>\n",
        "      <td> hom-alt</td>\n",
        "      <td> snp</td>\n",
        "      <td> snp</td>\n",
        "      <td> 1/1</td>\n",
        "      <td> 24</td>\n",
        "      <td> 21.04</td>\n",
        "      <td> 10</td>\n",
        "      <td> 1.000</td>\n",
        "      <td>  223,21,0</td>\n",
        "      <td> 21</td>\n",
        "      <td>  0,10</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>58771</th>\n",
        "      <th>T</th>\n",
        "      <th>C</th>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1</td>\n",
        "      <td> T</td>\n",
        "      <td> C</td>\n",
        "      <td> het-ref</td>\n",
        "      <td> ref</td>\n",
        "      <td> snp</td>\n",
        "      <td> 0/1</td>\n",
        "      <td> 21</td>\n",
        "      <td>    99</td>\n",
        "      <td> 52</td>\n",
        "      <td> 0.423</td>\n",
        "      <td> 402,0,341</td>\n",
        "      <td> 99</td>\n",
        "      <td> 30,22</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>60332</th>\n",
        "      <th>T</th>\n",
        "      <th>C</th>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1</td>\n",
        "      <td> T</td>\n",
        "      <td> C</td>\n",
        "      <td> het-ref</td>\n",
        "      <td> ref</td>\n",
        "      <td> snp</td>\n",
        "      <td> 0/1</td>\n",
        "      <td> 18</td>\n",
        "      <td>    99</td>\n",
        "      <td> 57</td>\n",
        "      <td> 0.456</td>\n",
        "      <td> 380,0,365</td>\n",
        "      <td> 99</td>\n",
        "      <td> 31,26</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 80,
       "text": [
        "                                       FORMAT  multiallele phase  GT1  GT2 a1  \\\n",
        "CHROM POS   REF ALT                                                             \n",
        "chr1  57856 T   A    GT:MQ:GQ:DP:VF:PL:GQX:AD            0     /    0    1  T   \n",
        "      57952 A   C    GT:MQ:GQ:DP:VF:PL:GQX:AD            0     /    1    1  C   \n",
        "      58211 A   G    GT:MQ:GQ:DP:VF:PL:GQX:AD            0     /    1    1  G   \n",
        "      58771 T   C    GT:MQ:GQ:DP:VF:PL:GQX:AD            0     /    0    1  T   \n",
        "      60332 T   C    GT:MQ:GQ:DP:VF:PL:GQX:AD            0     /    0    1  T   \n",
        "\n",
        "                    a2 zygosity vartype1 vartype2   GT  MQ     GQ  DP     VF  \\\n",
        "CHROM POS   REF ALT                                                            \n",
        "chr1  57856 T   A    A  het-ref      ref      snp  0/1  22     99  49  0.347   \n",
        "      57952 A   C    C  hom-alt      snp      snp  1/1  12   5.99  14  1.000   \n",
        "      58211 A   G    G  hom-alt      snp      snp  1/1  24  21.04  10  1.000   \n",
        "      58771 T   C    C  het-ref      ref      snp  0/1  21     99  52  0.423   \n",
        "      60332 T   C    C  het-ref      ref      snp  0/1  18     99  57  0.456   \n",
        "\n",
        "                            PL GQX     AD  \n",
        "CHROM POS   REF ALT                        \n",
        "chr1  57856 T   A    249,0,524  99  32,17  \n",
        "      57952 A   C       49,6,0   6   0,14  \n",
        "      58211 A   G     223,21,0  21   0,10  \n",
        "      58771 T   C    402,0,341  99  30,22  \n",
        "      60332 T   C    380,0,365  99  31,26  "
       ]
      }
     ],
     "prompt_number": 80
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "testing.zygosity.value_counts()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 81,
       "text": [
        "het-ref    26\n",
        "hom-alt    24\n",
        "dtype: int64"
       ]
      }
     ],
     "prompt_number": 81
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "testing = testing[testing.zygosity!='hom-ref']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 391
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "testing.vartype2.value_counts().plot(kind='barh')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 392,
       "text": [
        "<matplotlib.axes.AxesSubplot at 0x151efbe90>"
       ]
      },
      {
       "metadata": {},
       "output_type": "display_data",
       "png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD7CAYAAABkO19ZAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAD4tJREFUeJzt3X+M5Hddx/Hnwl2D9Lbe1RoshTgUEiNa7pqrh0prhjRi\nT1NaE4x/+M+BmihYakXEH9E7xWhTbDQaRQ2YrRBQiogSAcG437QhQtPQ3eNHW2jTJUChJd6dexds\nMXT84/Pd273tfrs3O5/vfOb7/j4fyWTmO7u3937123vv7Gvm5kCSJEmSJEmSJEmSJEmSJEkbzE37\nN9y/f/9oeXl52r+tJHXdMnBgJ7/wWZkH2dby8jKj0Sjs5ejRo8VnMJ/ZzBfvAuzf6d6d+qKPbmVl\npfQIrYqcL3I2MF+fueglKTgXfWZHjhwpPUKrIueLnA3M12dTfzIWGNV9kyTpPM3NzcEOd7aP6DOr\nqqr0CK2KnC9yNjBfn7noJSk4qxtJ6gCrG0lSIxd9ZtF7wsj5ImcD8/WZi16SgrOjl6QOsKOXJDUq\nsujn5uayXy666OISUZ4mek8YOV/kbGC+PttV5rfNX92cPl2ihZKk2Veko29j0cMcdv+SorKjlyQ1\nctFnFr0njJwvcjYwX5+56CUpODt6SeoAO3pJUiMXfWbRe8LI+SJnA/P12biL/hOtTCFJao0dvSR1\nwDQ7+jP19RCogDuB+4F3b/icW4HPAcvA23YylCQpn3EX/caHzAeAm4GXApcDrwC+C7gR+AFgP/DW\nDDN2SvSeMHK+yNnAfH02yZOx9wCPkpb/EvC9wCngCeCdwE8D/zvpgJKkyUzypmZPbrj9bWB3fX0I\nuBZ4DfAr9e1NjgCD+vZe0g8Hw/q4qq/HPa6P6u/qw+GwyPHafaV+f/Pt/Hg4HM7UPObrd76qqlhY\nWABgMBgwiXGL/dPAPGnLvgm4vr7/L4B7gfcDFwKPA98JPAxcsulr+GSsJI1pmk/Gjhpurx3PAx8i\nPRF7N3DLTobqsrXvyFFFzhc5G5ivz8atbi6qryvO7Uxu2nD75RPMI0nKzNfRS1IH+F43kqRGLvrM\noveEkfNFzgbm6zMXvSQFZ0cvSR1gRy9JauSizyx6Txg5X+RsYL4+c9FLUnB29JLUAZN09JO8qdkE\n8n9/mZ/fl/1rSlIERaqb0WiU/bK6eqJElKeJ3hNGzhc5G5ivz+zoJSm4Ih29XbokjcfX0UuSGrno\nM4veE0bOFzkbmK/PXPSSFJwdvSR1gB29JKmRiz6z6D1h5HyRs4H5+sxFL0nB2dFLUgfY0UuSGrno\nM4veE0bOFzkbmK/PXPSSFJwdvSR1gB29JKmRiz6z6D1h5HyRs4H5+sxFL0nB2dFLUgfY0UuSGrno\nM4veE0bOFzkbmK/PXPSSFJwdvSR1gB29JKmRiz6z6D1h5HyRs4H5+sxFL0nB2dFLUgdM0tHvyjvK\n+akHDmV+fh+rqydKjyFJT1OouhmFu5w+fRKI3xNGzhc5G5ivz+zoJSm4Ih19ehQczRw+9yCpLb6O\nXpLUyEWfWfSeMHK+yNnAfH3mopek4Ozos7Gjl9QeO3pJUiMXfWbRe8LI+SJnA/P12SSL/hjwpgk+\nLkmagkk6+qPAGeD2MT9uRy9JY5pmR/87wIPA3cD31fe9GPgIcC9w14b7JUkzYJxFfxD4WWA/8JPA\nD9X3/w1wE3AV8Gbgr3IO2DXRe8LI+SJnA/P12TjvXnkN8AHgifryr8BzgB8F7tzweRds/6WOAIP6\n9l7gADCsj6v6umvHydLSUvroMH187X++KMfR83ns8awcV1XFwsICAIPBgEmM0/fcDFxM6t4hde+n\ngF8Gnr/F59vRS1Im0+ro7wJuJD2KnweuB74JPAK8Zm0W4GU7GUSS1I5xFv19wD8Cy8CHgXtID81/\nDvh5YAn4LPDqDb+mdw9x1370iipyvsjZwHx9Nu6/MPVH9WWzw1vc9/vjjyNJys33usnGjl5Se3yv\nG0lSIxd9ZtF7wsj5ImcD8/WZi16SgrOjz8aOXlJ77OglSY1c9JlF7wkj54ucDczXZy56SQrOjj4b\nO3pJ7Zmkox/3b8ZmUuL7S7vm5/eVHkGStlSkuhmNRuEuq6sngPg9YeR8kbOB+frMjl6SgivS0dtl\nS9J4fB29JKmRiz6z6D1h5HyRs4H5+sxFL0nB2dFLUgfY0UuSGrnoM4veE0bOFzkbmK/PXPSSFJwd\nvSR1gB29JKmRiz6z6D1h5HyRs4H5+sxFL0nB2dFLUgfY0UuSGrnoM4veE0bOFzkbmK/PXPSSFJwd\nvSR1gB29JKmRiz6z6D1h5HyRs4H5+sxFL0nB2dFLUgfY0UuSGrnoM4veE0bOFzkbmK/PXPSSFJwd\nvSR1wCQd/a68o5yfemCpdfPz+1hdPVF6DKmoQtXNKPBlcQZmMN/a5fTpk5yv6B2v+frLjl6SgivS\n0adHW9I0zOFzQorA19FLkhq56LOrSg/Qsqr0AK2J3vGar79c9JIUnB29grOjVwx29JKkRi767KrS\nA7SsKj1Aa6J3vObrr7YW/RuBzwPvaunrS5LO0yQd/dqv3aoAvR+4Fnh0i4/Z0WuK7OgVwzQ7+gHw\nIHAH8Bngd4F7gGXgWP05fw1cDnwU+NWdDCVJymcn1c1LgL8EbgEuAw4BVwJXAdcAv0R6JD8E/izL\nlJ1SlR6gZVXpAVoTveM1X3/t5N0rv0R6FP8nwKuA++r7LyR9E7h7+y9xhPTDAcBe4ADp+wKsL5Ku\nHi/N2Dy5j7uWLy2A4XB49jbgscczf1xVFQsLCwAMBgMmMW7fMwA+BFxBWvRfAP52i897BDgIbPX+\nsHb0miI7esVQ6nX0/w68jvRIHlKN890TfD1JUgt2sujXHh59HHgP8F/AceBOYM+mz+mhqvQALatK\nD9Ca6B2v+fpr3I5+BXjZhuM/ry+bXb7TgSRJefleNwrOjl4x+F43kqRGLvrsqtIDtKwqPUBrone8\n5usvF70kBWdHr+Ds6BWDHb0kqZGLPruq9AAtq0oP0JroHa/5+stFL0nB2dErODt6xTBJR7+Td6/M\noMT3F/XR/Py+0iNIxRWpbkajUdjL4uJi8RnMt35ZXd3qDVS3Fr3jNV9/2dFLUnBFOvrRyM5Uksbh\n6+glSY1c9JlF7wkj54ucDczXZy56SQrOjl6SOsCOXpLUyEWfWfSeMHK+yNnAfH3mopek4OzoJakD\n7OglSY1c9JlF7wkj54ucDczXZy56SQrOjl6SOsCOXpLUyEWfWfSeMHK+yNnAfH3mopek4OzoJakD\n7OglSY1c9JlF7wkj54ucDczXZy56SQrOjl6SOsCOXpLUyEWfWfSeMHK+yNnAfH3mopek4OzoJakD\nJunod+Ud5fzUA0vSTJmf38fq6onSY2RX5BE9RH5EXwHDwjO0qSJuvoq42cB852OOWW0cfNWNJKmR\nj+gl6Swf0UuSOshFn11VeoCWVaUHaFFVeoCWVaUHaFlVeoCZ5aKXpODs6CXpLDt6SVIHueizq0oP\n0LKq9AAtqkoP0LKq9AAtq0oPMLNc9JIU3HZ9z4XA+4DLgGcDbwVuAxaA64HdwM8ADwLHgBfXl0vq\nz3vHFl/Tjl7SjOpnR38d8FXgAHAF8FHSlv4GcBB4O/DrGz7/B4FXAj8C/B5w6U6GkiTls92iPw78\nOHArcDWwWt//gfr608Cgvj0C/gV4EvhvYBE4lHHWjqhKD9CyqvQALapKD9CyqvQALatKDzCztnv3\nyi8CVwI/Bfwh8J/1/U/W19/e5ms8tfXdR1j//rCX9APDsD6u6uuuHi/N2Dy5j6Pn87jfx+kfMBkO\nh2dvA0WOq6piYWEBgMFgwCS263suBU4CT5CW/S+QtvJB4ARwFfA2Ul1zDLgB+GFgD+nR/suBr2/6\nmnb0kmZUzI5+u0f0V5AW+VPAt4DXA3du+PiI9a09IlU9i6QnY/+Apy95SdKU5fybsUeBM8Dt23xe\n8Ef0Fb7nd1dVxM0G5jsfMR/R534d/Wz+F5KkHvO9biTpLB/RS5I6yEWfXVV6gJZVpQdoUVV6gJZV\npQdoWVV6gJnlopek4OzoJeksO3pJUge56LOrSg/Qsqr0AC2qSg/Qsqr0AC2rSg8ws1z0khScHb0k\nnRWzo9/uvW5aUuL7iyQ9s/n5faVHaEWR6mY0GoW9LC4uFp/BfGYz384uq6snSqzE1tnRS1JwRTr6\n0Wg2OzBJmlW+jl6S1MhFn9naPwUWVeR8kbOB+frMRS9JwdnRS1IH2NFLkhq56DOL3hNGzhc5G5iv\nz1z0mS0tLZUeoVWR80XOBubrMxd9ZqdOnSo9Qqsi54ucDczXZy56SQrORZ/ZyspK6RFaFTlf5Gxg\nvj4r8fLKJWB/gd9XkrpsGThQeghJkiRJkiR11nXAA8AXgbcUniWXFeA4cB9wT33fxcDHgS8AHwP2\nFplsZ/4OeAz4zIb7ninPb5HO5wPAq6Y04yS2yncM+ArpHN4HHN7wsS7leyGwCHwO+Czwxvr+KOev\nKd8xYpy/5wCfIj2H+Xngj+v7O3X+ng08BAyA3aQw319yoEweIZ2IjW4DfqO+/Rbg1qlONJlrgCs5\ndxE25Xkp6TzuJp3Xh5j9V3Ftle8o8GtbfG7X8n0P60/U7QEeJP0Zi3L+mvJFOX8Az62vdwGfBK4m\n0/mbVvBD9SArwP8B/wDcMKXfu22bX7n0auCO+vYdwI3THWcidwMnN93XlOcG4L2k87lCOr+H2h9x\nIlvlg61ffda1fF8n/cEHOAPcD1xGnPPXlA9inD+Ab9bXF5AeHJ8k0/mb1qK/DPjyhuOvsH6SumwE\n/AdwL/CL9X3PI9UD1NfPKzBXTk15nk86j2u6fE5vIr107Z2s/2jc5XwD0k8unyLm+RuQ8n2yPo5y\n/p5F+mb2GOs1VZbzN61FH/V9iV9B+h/uMPAGUjWw0YhY2bfL08WsbwdeRKoFvgbc/gyf24V8e4B/\nAm4GTm/6WITztwd4PynfGWKdv6dIOV4A/Bjwyk0f3/H5m9ai/yrpyZQ1L+Tc70Zd9bX6+hvAP5N+\ndHqM1CcCXAo8XmCunJrybD6nL6jv65rHWf8D9A7Wf/ztYr7dpCX/LuCD9X2Rzt9avnezni/S+Vvz\nP8C/AQfp2PnbBTxM+pHrAmI8GftcYL6+fSHwCdIz37ex/qqi36RbT8ZCOkebn4zdKs/ak0EXkB5R\nPUyZv2k9rgHn5rt0w+1bgPfUt7uWbw74e+BPN90f5fw15Yty/i5hvXb6DuAu4Fo6eP4Ok54pf4j0\nsqCuexHpP/QS6eVea5kuJvX2XXx55XuBR4FvkZ5TeS3PnOe3SefzAeAnpjrpzmzO9zrS8jhO6ng/\nyLnPqXQp39WkH/2XWH+p4XXEOX9b5TtMnPN3BfBpUr7jwJvr+6OcP0mSJEmSJEmSJEmSJEmSJEmS\nJEnK4/8BQd3fHurNrxgAAAAASUVORK5CYII=\n",
       "text": [
        "<matplotlib.figure.Figure at 0x127c64ed0>"
       ]
      }
     ],
     "prompt_number": 392
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "testing.vartype2.value_counts().plot(kind='barh')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 388,
       "text": [
        "<matplotlib.axes.AxesSubplot at 0x125d107d0>"
       ]
      },
      {
       "metadata": {},
       "output_type": "display_data",
       "png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAD7CAYAAABjVUMJAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAEfFJREFUeJzt3X+QJPVZx/H3xoMg3OIdUhUIoRzwRwwCdxRIyhCsQUoC\nUX5opfQPU5WLGiv+IAQJRk3p7ZlSqSQYKpRGU4m1QAymwJgE0Zio1yFBEwJh78AEEOQsE35WAd6d\nCKhZ/+ged26Z3b2eb+9+v8/O+1U1NdM9vTOf7dnrZ+Z5ZuZAkiRJkiRJkiRJkiRJkiRJAU3lDjCw\nZcuW+V27duWOIUnR7AK2jvODL+k4yNh27drF/Px82NP27duzZ5jU/JGzmz//KXp+YMu4x91iCkB0\ne/bsyR0hSeT8kbOD+XOLnj+FBUCSJpQFoCPbtm3LHSFJ5PyRs4P5c4ueP0UxQ2BgvulnSZIO0tTU\nFIx5LPcVQEeqqsodIUnk/JGzg/lzi54/hQVAkiaULSBJCswWkCSpNQtAR6L3ESPnj5wdzJ9b9Pwp\niioARx55VO4IkjQxipoBADgHkKSD5wxAktSaBaAj0fuIkfNHzg7mzy16/hQWAEmaUM4AJCkwZwCS\npNZWKgC3t7y9PnBLB9uEE72PGDl/5Oxg/tyi50+xUgE4a01SSJLW3Ep9o/3ARupn7TPAk8DJwF3A\nG5ttzgfeDzwLfBE4EbgQOAK4FvgB4JDm5z/d3NYVzTbDnAFIUkurOQMYPhpvBS4DTqI+yL8GOAz4\nEPDjwOnAMUM/8y7g74FXAz8CvBc4fJyQkqTubWix7R3AI83lOeAE6mf9DwMPNes/CvxCc/k86mf5\n72iWXwocv9KdzMzMALBp0ya2bt1Kv98HFvp0pS5fc801ofKup/zDPdwS8pi/rHzrLX9VVczOzgLQ\n6/VYTfua8z4HDm6vBd5E/b/Rf35o/UVD290JfO+I21x8WwPz1F8JHdLOnTtzR0gSOX/k7PPz5s8t\nen4O7NS0slLfaB8wzYv79tcCXwE+DjwAnAP8K3Aj9czgQuB3gSOBS5ufOQ24e8RtDTgDkKSW1moG\nMOrI/Dx1y+dW6sHw40PbvZt6+LsbuBfYMXQ7HuUlKbOVCsCRzXlF3d4ZuBS4vrn8t8CrqIfAbx/a\n7jngrcCp1O8cGqz//KLbWheG+4gRRc4fOTuYP7fo+VP4SWBJmlB+F5AkBeZ3AUmSWrMAdCR6HzFy\n/sjZwfy5Rc+fwgIgSRPKGYAkBbZuZgDT05tzR5CkiVFUAdi796ncEcYWvY8YOX/k7GD+3KLnT1FU\nAZAkrZ2iZgD2/yWpnXUzA5AkrR0LQEei9xEj54+cHcyfW/T8KSwAkjShnAFIUmDOACRJrVkAOhK9\njxg5f+TsYP7coudPYQGQpAnlDECSAnMGIElqzQLQkeh9xMj5I2cH8+cWPX8KC4AkTShnAJIUmDMA\nSVJrFoCORO8jRs4fOTuYP7fo+VNYACRpQjkDkKTAnAFIklqzAHQkeh8xcv7I2cH8uUXPn2JD7gDD\nmpcyRZme3hz6P6uXpKWUdMSdhxJnAFM4m5BUKmcAkqTWLAAdid5HjJw/cnYwf27R86ewAEjShHIG\nsCJnAJLK5QxAktRaVwXg9o5uJ6zofcTI+SNnB/PnFj1/iq4KwFkd3Y4kaY10NQPYD2wE+sAM8CRw\nMnAX8MZmm6uAC4H/AT4LXLnoNpwBSFJLKTOArj4JPHyE3AqcBDxK3Ro6C7gPuAT4/mabIzu6X0nS\nmFZjCHwH8Ah1UZgDvgt4BngO+AjwE8B/rcL9ZhW9jxg5f+TsYP7coudPsRrfBfT80OX/BQ5pzs8E\nzgXeAPxKc3mRbUCvubyJ+sVEv1mumvO1Xm6Wmj+Sfr8/cnlubm7Z60tfjp7fZZcnZbmqKmZnZwHo\n9Xqk6GoGsA+Ypj56XkHd6we4FrgTuBk4AngC+A7gIeDoRbfhDECSWiptBrD4aDlPXRw+BRxGHfTy\nju5XkjSmrmYAg6FuBVw0tP5S4HrgMeDVwBbgVOCGju63GIOXaFFFzh85O5g/t+j5U/hJYEmaUH4X\n0IqcAUgql98FJElqzQLQkeh9xMj5I2cH8+cWPX8KC4AkTShnACtyBiCpXM4AJEmtFVYApoo7TU9v\nPqjk0fuIkfNHzg7mzy16/hSr8V1AY7PVIklrp6gZgAVAktpxBiBJas0C0JHofcTI+SNnB/PnFj1/\nCguAJE0oZwCSFJgzAElSaxaAjkTvI0bOHzk7mD+36PlTWAAkaUI5A5CkwJwBSJJaswB0JHofMXL+\nyNnB/LlFz5/CAiBJE8oZgCQF5gxAktSaBaAj0fuIkfNHzg7mzy16/hQWAEmaUM4AJCkwZwCSpNYs\nAB2J3keMnD9ydjB/btHzp7AASNKEKmoGkDvAUqanN7N371O5Y0jSi6TMAAorAKXWgCkcUEsqkUPg\nAkTvI0bOHzk7mD+36PlTWAAkaULZAjootoAklckWkCSpNQtAR6L3ESPnj5wdzJ9b9PwpVqMAzABX\nJFwvSVoDqzED2A7sB65ueb0zAElqqYQZwLuA+4EvAK9s1n038DfAncBtQ+slSQXoogCcDvw0sAV4\nPfCDzfo/AS4FzgCuBP6og/sqVvQ+YuT8kbOD+XOLnj/Fhg5u42zgE8BzzenTwGHAa4CbhrY7dOWb\n2gb0msubgK1Av1mumvM8y4M/kn5/9PLc3Nyy15e+HD2/yy5PynJVVczOzgLQ6/VI0cUM4DLgKOre\nPtS9/WeAXwRePmJ7ZwCS1JHcM4DbgEuon/VPAxcCzwIPA29otpkCTu3gviRJHemiANwNfBzYBfw1\ncAf1U/mfAX4OmAPuBS4a+pl193R68BItqsj5I2cH8+cWPX+KLmYAAL/XnBa7YMS6HR3dpyQpgd8F\ndFCcAUgqU+4ZgCQpIAtAR6L3ESPnj5wdzJ9b9PwpLACSNKGcARwUZwCSyuQMQJLUWmEFYKrI0/T0\n5hWTR+8jRs4fOTuYP7fo+VN09TmATthmkaS1U9QMwAIgSe04A5AktWYB6Ej0PmLk/JGzg/lzi54/\nhQVAkiaUMwBJCswZgCSpNQtAR6L3ESPnj5wdzJ9b9PwpLACSNKGcAUhSYM4AJEmtWQA6Er2PGDl/\n5Oxg/tyi509hAZCkCeUMQJICcwYgSWrNAtCR6H3EyPkjZwfz5xY9fwoLgCRNKGcAkhSYMwBJUmsW\ngI5E7yNGzh85O5g/t+j5U1gAJGlCOQOQpMBSZgAbuo2SpvlFJACmpzezd+9TuWNI61ZhLaD5wKed\nBWRYX/n37XuagxG9h2v+vKLnT1FYAZAkrZWSei7z9TM/aWAK50LS8vwcgCSpNQtAZ6rcARJVuQOM\nLXoP1/x5Rc+fYq0LwNuArwE3rPH9SpIWWY0ZwOA2RzVvvw6cCzwy4jpnAFrEGYC0khJmAD3gfuA6\n4B7gt4A7gF3ATLPNHwMnAp8B3t7R/UqSxtRlC+h7gD8ELgeOA84ETgPOAM4G3kr9zL8PXNPh/Rai\nyh0gUZU7wNii93DNn1f0/Cm6LAD/Rv2s/3XAecDdwF3A91EXB0lSQbqaAfSAW4BTgPcBDwAfGrHd\nw8DpwKjP98/Dm5qbAtgEbKV+wQALz1Bdnpzlc/5/BjB4ltbv9112eaKXq6pidnYWgF6vx44dO2DM\nY/lqFIAfBd5NPez9T+p20AvAk6xYABz4aZhDYGklJQyBYeHo/TngY8A/AbuBm4CNi7ZZh6rcARJV\nuQOMLXoP1/x5Rc+foqtvA90DnDq0/IHmtNiJHd2fJCmR3wWkgtkCklZSSgtIkhSIBaAzVe4Aiarc\nAcYWvYdr/ryi509hAZCkCeUMQAVzBiCtxBmAJKk1C0BnqtwBElW5A4wteg/X/HlFz5+iq88BdKSk\njpRym57enDuCtK6VdMSdt98rSe04A5AktWYB6Ej0PmLk/JGzg/lzi54/hQVAkiaUMwBJCswZgCSp\nNQtAR6L3ESPnj5wdzJ9b9PwpLACSNKGcAUhSYM4AJEmtWQA6Er2PGDl/5Oxg/tyi509hAZCkCeUM\nQJICcwYgSWrNAtCR6H3EyPkjZwfz5xY9fwoLgCRNKGcAkhSYMwBJUmsWgI5E7yNGzh85O5g/t+j5\nU1gAJGlCOQOQpMBSZgAbuo2SpvlFJCm76enN7N37VO4Yq6qwFtB84NPOAjJMav7I2c2f/zQ6/759\nT7PeFVYAJElrpaSey3xdeSWpBFNEmEv6OQBJUmsWgM5UuQMkqnIHSFDlDpCoyh0gUZU7QKIqd4Bs\nLACSNKGcAUjSSM4AJEnr1LgF4AjgVmAOuAf4KWAPMAPcBewGXtlsOwPcAPwj8ADw8+OGLVuVO0Ci\nKneABFXuAImq3AESVbkDJKpyB8hm3AJwPvBNYCtwCvAZ6v7Nk8DpwAeBdwxtfzJwDvBDwG8Dx455\nv5Kkjoz7VRC7gfcBVwF/BXyxWf+J5vyrwE82l+eBTwHPN6edwJnNukW2Ab3m8ibq+tJvlqvmvNTl\nwbpS8rRdHqwrJU+b5X5hedou9wvL03a5X1ietsv9Za5vlppvDO33+9mXq6pidnYWgF6vR4qUIfAm\n4MeAtwD/ALyZ+tn/U8AZwHupn/Vvb+5npvm564CbgVsW3Z5DYEkFcQi8lGOB54A/oz7Qn7bMtlPA\nxcBLge+kLrFfGfN+C1blDpCoyh0gQZU7QKIqd4BEVe4AiarcAbIZtwV0CvWB/1vAC8AvATcNXT/4\nRqXB5d3UrZ+jgd8BHhvzfiVJHVmLzwFsB/YDV6+wnS0gSQWxBdSV8veiJE2YtSgAO4A/WIP7yazK\nHSBRlTtAgip3gERV7gCJqtwBElW5A2TjJ4ElaUL5XUCSNJIzAEnSOmUB6EyVO0CiKneABFXuAImq\n3AESVbkDJKpyB8jGAiBJE2rcD4KtkpJGEpIm2fT05twRVl1RBSDCwEWS1gtbQB0ZfFtfVJHzR84O\n5s8tev4UFgBJmlAlNd3nbQFJUjt+DkCS1JoFoCPR+4iR80fODubPLXr+FBaAjszNzeWOkCRy/sjZ\nwfy5Rc+fwgLQkWeeeSZ3hCSR80fODubPLXr+FBYASZpQFoCO7NmzJ3eEJJHzR84O5s8tev4UJb0N\ndA7YkjuEJAWzC9iaO4QkSZIkSZK0hPOB+4B/Ad6ZOcvB2gPsBu4G7mjWHQV8DngA+CywKUuy0f4U\neBy4Z2jdcnl/g/rxuA84b40yLmdU/hngG9SPwd3ABUPXlZT/eGAn8M/AvcDbmvVR9v9S+WeIsf8P\nA75MPWP8GvD7zfoo+3+p/DPE2P/L+jbgQaAHHEL9S74qZ6CD9DD1H9Cw9wC/1lx+J3DVmiZa3tnA\naRx4AF0q70nUj8Mh1I/Lg+R/t9io/NuBXx2xbWn5j2FhQLcRuJ/6bzzK/l8qf5T9D3B4c74B+BLw\nWuLsfxidv5P9n/sXO5M64B7gv4E/By7OGaiFxe+gugi4rrl8HXDJ2sZZ1heApxetWyrvxcCN1I/H\nHurH58zVj7isUflh9LvYSsv/GPU/SID9wNeB44iz/5fKDzH2P8Czzfmh1E86nybO/ofR+aGD/Z+7\nABwH/PvQ8jdY+OMq2Tzwd8CdwFuadS+jblPQnL8sQ642lsr7curHYaDkx+RS6rfAfYSFl/Al5+9R\nv5L5MjH3f486/5ea5Sj7/yXURexxFtpZkfb/qPzQwf7PXQCifv/zWdT/EC4Afpm6RTFsnli/20p5\nS/xdPgicQN2eeBS4epltS8i/EfgL4DJg36LrIuz/jcDN1Pn3E2v/f4s65yuAHwbOWXR96ft/cf4+\nHe3/3AXgm9RDpoHjObB6lerR5vxJ4C+pX2I9Tt0vBTgWeCJDrjaWyrv4MXlFs640T7DwD/fDLLzM\nLTH/IdQH/xuATzbrIu3/Qf6PspA/0v4f+A/gVuB0Yu3/gUH+M4i5/19kA/AQ9UvLQ4kxBD4cmG4u\nHwHcTj1pfw8L72L6dcoaAkO9jxcPgUflHQyRDqV+hvEQZXxivMeB+Y8dunw58LHmcmn5p4Drgfcv\nWh9l/y+VP8r+P5qF9si3A7cB5xJn/y+V/5ihbUre/yu6gPqdBQ9Sv32pdCdQ7+A56rfFDTIfRT0X\nKPFtoDcCjwAvUM9c3szyeX+T+vG4D3jdmiYdbXH+n6U+KO2m7oF+kgNnLiXlfy31S/g5Ft6ydz5x\n9v+o/BcQZ/+fAnyVOv9u4MpmfZT9v1T+KPtfkiRJkiRJkiRJkiRJkiRJkiRJkpTi/wBSCrz6bgFU\nQAAAAABJRU5ErkJggg==\n",
       "text": [
        "<matplotlib.figure.Figure at 0x151c827d0>"
       ]
      }
     ],
     "prompt_number": 388
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Kunal's Framework"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#KUNAL'S FRAMEWORK\n",
      "class Vcf(object):\n",
      "    def __init__(self, filename, sample_col, generate_features=False,\n",
      "                 autosome_only=False, no_halfs=False):\n",
      "\n",
      "        # Load VCF\n",
      "        self._variant_df = self._load_vcf(filename)\n",
      "        \n",
      "        # Sample Column Name\n",
      "        self.sample_col = sample_col\n",
      "\n",
      "        \n",
      "        # Filter\n",
      "        if autosome_only:\n",
      "            self._variant_df = self._variant_df[self._autosome_variants]\n",
      "            \n",
      "        if no_halfs:\n",
      "            self._variant_df = self._variant_df[self._not_half_calls]\n",
      "            \n",
      "        #remove haploid called variants, occurs with sex chromosomes, mitochondria, or some structural variants    \n",
      "        self._variant_df = self._variant_df[self._not_haploid]\n",
      " \n",
      "        # Process variants\n",
      "        self._process()\n",
      " \n",
      "        # Generate features\n",
      "        if generate_features:\n",
      "            self._generate_features()\n",
      "        else:\n",
      "            self._features_df = pd.DataFrame([])\n",
      "            \n",
      "\n",
      "\n",
      "    def _process(self):\n",
      "        df = self._variant_df\n",
      "\n",
      "#         # Ignore CNV and Repeats\n",
      "#         discard_vars = ['<INS:ME:ALU>', '<INS:ME:L1>',\n",
      "#                         '<CGA_CNVWIN>', '<INS:ME:SVA>']\n",
      "#         df = df[~df.ALT.isin(discard_vars)]\n",
      "\n",
      "        # Parse many fields\n",
      "        df = self._parse(df)\n",
      "        \n",
      "        df = df[df['phase'] != '-']\n",
      "\n",
      "        # Ignore sex chromosomes with - as phaser\n",
      "        df = df[df.phase != '-']\n",
      "\n",
      "        # Ignore structural variants\n",
      "        df = df[(df.vartype1 != 'sv') & (df.vartype2 != 'sv')]\n",
      "\n",
      "        # Fill missing allelic depth with 0,0 reads\n",
      "        #df['AD'].fillna(value='0,0', inplace=True)\n",
      "\n",
      "        self._variant_df = df\n",
      "\n",
      "\n",
      "    def _get_allele(self, line, gt_col):\n",
      "        '''\n",
      "        Returns allelic base, handles multi-allelic variants\n",
      "        '''\n",
      "        alleles = [line['REF']]\n",
      "        alleles.extend(line['ALT'].split(\",\"))\n",
      "        a1 = \".\"\n",
      "        try:\n",
      "            a1 = alleles[int(line[gt_col])]  #returns missing if gt_int_call is \".\"\n",
      "        except:\n",
      "            a1 = \".\"\n",
      "        return a1\n",
      "        \n",
      "        \n",
      "    def _get_GT_multisample_vcf(self, line, sample_col, gt_index):\n",
      "        '''\n",
      "        Slow parser for multisample vcf\n",
      "        '''\n",
      "        return int( line[sample_col].split(line['phase'])[int(gt_index)])\n",
      "\n",
      "\n",
      "\n",
      "    def _get_allele_bases(self, df, sample_col, single_sample_vcf=True):\n",
      "        '''\n",
      "        Adds phase, GT1, GT2, a1, a2 to self._variants_df dataframe\n",
      "        \n",
      "        phase : { /, | }, unphased or phased call\n",
      "        GT1: int, first allele call in the numeric sample genotype column\n",
      "        GT2: int, second allele call in the numeric sample genotype column\n",
      "        a1: {A, T, G, C, AA, etc}, nucleotide base representation for GT1\n",
      "        a2: {A, T, G, C, AA, etc}, nucleotide base representation for GT2\n",
      "        \n",
      "        '''\n",
      "        \n",
      "        \n",
      "        if single_sample_vcf:\n",
      "            df['phase'] = df[sample_col].str[1]\n",
      "            df['GT1'] = df[sample_col].str[0]\n",
      "            df['GT1'] = df['GT1'].astype(int)\n",
      "            df['GT2'] = df[sample_col].str[2]\n",
      "            df['GT2'] = df['GT2'].astype(int)\n",
      "        \n",
      "        \n",
      "        if not single_sample_vcf:\n",
      "            df['phase'] = df.apply(get_phase, args=['GT'], axis=1)  #get phase\n",
      "            df = df[df.phase != \"-\"]  #likley occurs at sex chromosome sites\n",
      "            df['GT1'] = df.apply(self._get_GT_multisample, args=[sample_col, 0], axis=1)\n",
      "            df['GT2'] = df.apply(self._get_GT_multisample, args=[sample_col, 1], axis=1)\n",
      "            \n",
      "            \n",
      "        \n",
      "        #SLOW PROCESS MULTIPLE ALLELE GENOTYPES\n",
      "        df_multi = df[(df.GT1.astype(int)>1) | (df.GT2.astype(int)>1)] #select all multi-alleleic variants\n",
      "        df_multi['a1'] = df_multi.apply(self._get_allele, args=['GT1'], axis=1)  #\n",
      "        df_multi['a2'] = df_multi.apply(self._get_allele, args=['GT2'], axis=1)\n",
      "        \n",
      "        \n",
      "        #FAST PROCESS SIMPLE ALLELE GENOTYPES\n",
      "        df_simple = df[~df.index.isin(df_multi.index)][['REF', 'ALT', 'GT1', 'GT2']]  #dropping multiallele variants, minimize memory usage\n",
      "        \n",
      "        df_gt1_ref = df_simple[df_simple.GT1==0][['REF']]  #get a1 ref alleles\n",
      "        df_gt1_ref.columns = ['a1']\n",
      "        df_gt2_ref = df_simple[df_simple.GT2==0][['REF']]  #get a2 ref alleles\n",
      "        df_gt2_ref.columns = ['a2']\n",
      "        \n",
      "        \n",
      "        df_gt1_alt = df_simple[df_simple.GT1==1][['ALT']]  #get a1 alt alleles\n",
      "        df_gt1_alt.columns = ['a1']\n",
      "        df_gt2_alt = df_simple[df_simple.GT2==1][['ALT']]  #get a2 alt alleles\n",
      "        df_gt2_alt.columns = ['a2']\n",
      "        \n",
      "        \n",
      "        gt1_alleles = pd.concat([df_gt1_ref,df_gt1_alt])  #merging GT1 allele bases into a single df\n",
      " \n",
      "        gt2_alleles = pd.concat([df_gt2_ref,df_gt2_alt])  #merging GT2 allele bases into a single df\n",
      "\n",
      "        gt1_2_allele_df = gt1_alleles.join(gt2_alleles, how='outer')  #Joining the GT1 and GT2 simple allele bases \n",
      "        \n",
      "        \n",
      "        df = df.join(gt1_2_allele_df, how='inner')  #Adding simle allele a1 and a2 columns to original df\n",
      "        df = df.append(df_multi)  #Adding multi-alleleic bases to original df\n",
      "        \n",
      "        return df\n",
      "\n",
      "\n",
      "\n",
      "    def _zygosity_fast(self, df):\n",
      "        '''\n",
      "        This function quickly assigns zygosity states to \n",
      "        each variant using set logic.\n",
      "        \n",
      "        zygosity: {hom-ref, hom-miss, het-miss, het-alt, hom-alt, het-ref}\n",
      "        examples:  REF     ALT\n",
      "        hom-ref    A        A/A\n",
      "        hom-miss   A        ./.\n",
      "        het-miss   A        A/.\n",
      "        het-alt    A        T/G\n",
      "        hom-alt    A        T/T\n",
      "        het-ref    A        A/T\n",
      "        '''\n",
      "        \n",
      "        #A/A\n",
      "        df_hom_ref = df[ (df['a1'] == df['REF']) & (df['a2'] == df['REF'])]\n",
      "        df_hom_ref['zygosity'] = 'hom-ref'\n",
      "        \n",
      "        #./.\n",
      "        df_hom_miss = df[ (df['a1'] == '.') & (df['a2'] == '.')]\n",
      "        df_hom_miss['zygosity'] = 'hom-miss'\n",
      "        \n",
      "        #A/.\n",
      "        df_het_miss = df[ (df['a1'] == '.') | (df['a2'] == '.') ]\n",
      "        df_het_miss['zygosity'] = 'het-miss'\n",
      "        \n",
      "        # NO MISSING CALLS\n",
      "        df_not_miss = df[~df.index.isin( set(df_hom_miss.index) | set(df_het_miss.index))  ]\n",
      "        \n",
      "        #T/G\n",
      "        df_het_alt = df_not_miss[ ((df_not_miss['a1'] != df_not_miss['REF']) & (df_not_miss['a2'] != df_not_miss['REF'])) & (df_not_miss['a1'] != df_not_miss['a2']) ]\n",
      "        df_het_alt['zygosity'] = 'het-alt'\n",
      "        \n",
      "        #T/T\n",
      "        df_hom_alt = df_not_miss[ (((df_not_miss['a1'] != df_not_miss['REF']) & (df_not_miss['a2'] != df_not_miss['REF']))) & (df_not_miss['a1'] == df_not_miss['a2']) ]\n",
      "        df_hom_alt['zygosity'] = 'hom-alt'\n",
      "        \n",
      "        #A/T\n",
      "        df_het_ref = df_not_miss[ ((df_not_miss['a1'] == df_not_miss['REF']) & (df_not_miss['a2'] != df_not_miss['REF'])) | ((df_not_miss['a1'] != df_not_miss['REF']) & (df_not_miss['a2'] == df_not_miss['REF']))  ]\n",
      "        df_het_ref['zygosity'] = 'het-ref'\n",
      "        \n",
      "        #Appending dataframes\n",
      "        df_zygosity = pd.concat([df_hom_ref, df_hom_miss, df_het_miss, df_het_ref, df_het_alt, df_hom_alt])\n",
      "        \n",
      "        \n",
      "        \n",
      "        #print set(df_zygosity.index) - set(df.index)\n",
      "        assert len(df_zygosity) == len(df) #ensures no variants were missed or double counted\n",
      "        return df_zygosity\n",
      "\n",
      "\n",
      "\n",
      "    def _vartype_map(self, ref_alt_bases):\n",
      "        '''\n",
      "        This function assigns the following vartypes to the \n",
      "        allele specified by allele_base_col: snp, mnp, ins, del, indel or SV\n",
      "        \n",
      "        '''\n",
      "        ref, alt = ref_alt_bases\n",
      "        len_diff = len(ref) - len(alt)\n",
      "        \n",
      "        if ref == alt: return 'ref' #Orderd by frequency of the variant to reduce complexity\n",
      "        \n",
      "        if len_diff == 0:\n",
      "            base_diff = [nt for i,nt in enumerate(alt) if ref[i] != alt[i]]\n",
      "            if len(base_diff) == 1: return 'snp' \n",
      "            else: return 'mnp'\n",
      "        \n",
      "        if len_diff > 0:\n",
      "            base_diff = [nt for i,nt in enumerate(alt) if ref[i] != alt[i]]\n",
      "            if len(base_diff) > 0: return 'indel'\n",
      "            else: return 'del'\n",
      "            \n",
      "        if len_diff < 0:\n",
      "            return 'ins'\n",
      "        \n",
      "        elif is_sv(ref,alt): return 'sv'\n",
      "        \n",
      "        else: return 'indel or SV'\n",
      "\n",
      "\n",
      "    def _parse(self, df):\n",
      "        '''\n",
      "        Main processing function. Initial processing steps that might include\n",
      "        feature generation or alternate representation of the dataframe.\n",
      "\n",
      "        '''\n",
      "\n",
      "        # Phase genotype and remove edge cases, likely at sex chromosome sites\n",
      "        # series['phase'] = self._get_phase(series['GT'])\n",
      "        # if series['phase'] == '-':\n",
      "        #    return series\n",
      "\n",
      "        # Get allele base sequences\n",
      "        df = self._get_allele_bases(df, self.sample_col)  #adding 'phase', 'GT1', 'GT2', 'a1', 'a2' to df\n",
      "        \n",
      "        # Zygosity: hom-ref, het-ref, hom-alt, het-alt\n",
      "        df = self._zygosity_fast(df)\n",
      "        \n",
      "        # Set the variant type based on number of bases in reference vs allele\n",
      "        df['vartype1'] = map(self._vartype_map, df[['REF','a1']].values)\n",
      "        df['vartype2'] = map(self._vartype_map, df[['REF','a2']].values)\n",
      "        \n",
      "        # Multiallelic site if more than one alternative allele\n",
      "        df['multiallele'] = df.ALT.map(lambda x: 1 if \",\" in x else 0)\n",
      "        \n",
      "        return df\n",
      "\n",
      "\n",
      "    \n",
      "        "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 47
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def __aaf(INFO_dict,alt_all_num):\n",
      "    AN = float(INFO_dict[\"AN\"])\n",
      "    AC_raw = INFO_dict[\"AC\"].split(\",\")\n",
      "    AC = AC_raw[int(alt_all_num)-1]  #correcting for reference allele\n",
      "    return float(AC) / AN"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 44
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def INFO(info_field,alt_all_num):\n",
      "    info_data = info_field.split(\";\")\n",
      "    INFO_dict = dict()\n",
      "    ordered_keys = []\n",
      "    for items in info_data:\n",
      "        #print items\n",
      "        k,v = items.split(\"=\")\n",
      "        ordered_keys.append(k)\n",
      "        if k == \"EFF\":\n",
      "            annot = [i for i in v.split(\",\") if i[-2] == str(alt_all_num)]\n",
      "            snpeff_df = __snpeff(annot,alt_all_num)\n",
      "            INFO_dict[k] = snpeff_df\n",
      "        else: INFO_dict[k] = v\n",
      "    INFO_dict['aaf'] = __aaf(INFO_dict,alt_all_num)\n",
      "    return INFO_dict"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 45
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def __format_parsing(sample_Series,format_field):\n",
      "    lol = [s.split(\":\") for s in sample_Series]\n",
      "    sample_DataFrame = pd.DataFrame(lol)\n",
      "    sample_DataFrame.columns = format_field\n",
      "    sample_DataFrame.index = sample_Series.index\n",
      "    return sample_DataFrame\n",
      "\n",
      "def __sample_parsing(sample_fields, sample_ids, format_field, alt_all_num):\n",
      "    hets_df, hom_df = [], []\n",
      "    sample_Series = pd.Series(sample_fields)\n",
      "    sample_Series.index=sample_ids\n",
      "    homozygous_call = str(alt_all_num) + \"/\" + str(alt_all_num)\n",
      "    \n",
      "    hets = sample_Series[[True if str(alt_all_num) in i.split(\":\",2)[0] and i.split(\":\",2)[0] != homozygous_call else False for i in sample_Series]]\n",
      "    if len(hets) != 0: \n",
      "        hets_df = __format_parsing(hets,format_field) #dataframe with Format info for each het minor allele sample\n",
      "    \n",
      "    hom = sample_Series[[True if i.split(\":\",2)[0] == homozygous_call else False for i in sample_Series]]\n",
      "    if len(hom) != 0: \n",
      "        hom_df = __format_parsing(hom,format_field)  #dataframe with Format info for each hom minor allele sample\n",
      "    \n",
      "    return hets_df, hom_df\n",
      "    \n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 46
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "TODO: Make Header Class"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "class Vcf_metadata(object):\n",
      "    \n",
      "    def __init__(self, filename):\n",
      "        \n",
      "        if filename.endswith('.gz'):\n",
      "            self.compression = 'gzip'\n",
      "            if filename+'.tbi' in os.listdir(os.path.split(filename)[0]):\n",
      "                header_lines = os.popen('tabix -H ' + filename).readlines()\n",
      "                self.header = [l.replace('#CHROM','CHROM') for l in header_lines if l.startswith('#')]\n",
      "            os.system('tabix -p vcf ' + filename)\n",
      "            header_lines = os.popen('tabix -H ' + filename).readlines()\n",
      "            self.header = [l.replace('#CHROM','CHROM') for l in header_lines if l.startswith('#')]\n",
      "        \n",
      "        else:\n",
      "            self.compression = ''\n",
      "            header_lines = os.popen('head -5000 ' + filename).readlines()\n",
      "            self.header = [l.replace('#CHROM','CHROM') for l in header_lines if l.startswith('#')]\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# def get_vcf_df(file_path, sample_name):\n",
      "    \n",
      "#     def get_header(vcf_path):\n",
      "#         if vcf_path.endswith('.gz'):\n",
      "#             return os.popen('tabix -H ' + vcf_path).readlines()\n",
      "#         header_lines = os.popen('head -5000 ' + vcf_path).readlines()\n",
      "#         return [l for l in header_lines if l.startswith('#')]\n",
      "    \n",
      "    \n",
      "    \n",
      "#     header = get_header(file_path)\n",
      "#     df = pd.read_table(file_path, sep=\"\\t\", compression='gzip', skiprows=(len(header)-1))\n",
      "#     df.set_index(['#CHROM', 'POS', 'REF', 'ALT'], inplace=True, drop=False)\n",
      "\n",
      "#     return df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "class vcf(object):\n",
      "    '''\n",
      "    Loads in a vcf file, aware of gzipped files.\n",
      "    '''\n",
      "    \n",
      "    def __init__(self, filename):\n",
      "        \n",
      "        #Header\n",
      "        header_parsed = Vcf_metadata(filename)\n",
      "        self.header = self.get_header_df(header_parsed.header)  #header parsed into key/values dataframe\n",
      "    \n",
      "        #Vcf df indexed on CHROM, POS, REF, ALT\n",
      "        self.df = pd.read_table(filename, sep=\"\\t\", compression=header_parsed.compression, skiprows=(len(self.header)-1))\n",
      "        self.df.columns = header_parsed.header[-1].rstrip('\\n').split(\"\\t\")\n",
      "        self.df.set_index(['CHROM', 'POS', 'REF', 'ALT'], inplace=True, drop=False)\n",
      "        self.df_bytes = self.df.values.nbytes + self.df.index.nbytes + self.df.columns.nbytes\n",
      "        \n",
      "        #Sample Ids\n",
      "        self.samples = self.df.columns[8:]\n",
      "        \n",
      "        #Format values\n",
      "        self.FORMAT = self.df[self.df.columns[8]]\n",
      "        \n",
      "        \n",
      "    \n",
      "    \n",
      "    def get_header_df(self, header_txt):\n",
      "        '''\n",
      "        Parses header into pandas DataFrame\n",
      "        '''\n",
      "        key_value_header = [i.replace('##','').replace('\\n','').split('=',1) for i in header_txt if '##' in i]\n",
      "        key_value_header.append(['SampleIDs',header_txt[-1].rstrip('\\n').split('\\t')[9:]])\n",
      "        header_df =  pd.DataFrame.from_records(key_value_header)\n",
      "        header_df.set_index(0,inplace=True)\n",
      "        header_df.index.name = 'header_keys'\n",
      "        header_df.columns = ['header_values']\n",
      "        return header_df\n",
      "    \n",
      "    \n",
      "    def get_vcf_samples_df(self, split_cols):\n",
      "        '''\n",
      "        This function creates a samples_df containing sample\n",
      "        level information for all non-missing variant calls\n",
      "        indexed on CHROM, POS, REF, ALT\n",
      "        \n",
      "        vcf_df\n",
      "        \n",
      "        split_cols\n",
      "        \n",
      "        '''\n",
      "            \n",
      "        genotypes = self.df[self.samples].groupby(by=self.FORMAT)  #genotypes grouped by FORMAT variant annotations\n",
      "        \n",
      "        #Iterate through genotype groups, dropping missing calls\n",
      "        master_df = []\n",
      "        for name,group in genotypes:\n",
      "            temp_group = genotypes.get_group(name)  #group of interest\n",
      "            del temp_group['FORMAT']  #remove the format column\n",
      "            temp_group.replace(to_replace='.', value=nan, inplace=True)  #replace . with none, allows stack to remove null columns, space savings\n",
      "            \n",
      "            temp_group = temp_group.stack()  #stacking samples for each variant\n",
      "        \n",
      "            #creating sample dataframe\n",
      "            temp_group_data = pd.DataFrame(temp_group.str.split(':').tolist())\n",
      "            temp_group_data.index = temp_group.index\n",
      "            temp_group_data.columns = name.split(':')\n",
      "            temp_group_data.replace(to_replace='.', value=nan, inplace=True)\n",
      "            \n",
      "            master_df.append(temp_group_data)\n",
      "        \n",
      "        #Concatenating all genotype groups\n",
      "        self.sample_df = pd.concat(master_df)\n",
      "        self.sample_df.index.names = ['CHROM', 'POS', 'REF', 'ALT', 'SAMPLE_ID']\n",
      "        \n",
      "        #spliting user-defined columns\n",
      "        for col in split_cols:\n",
      "            for i in range(0, split_cols[col]):\n",
      "                self.sample_df[col + '_' + str(i)] = self.sample_df[col].str.split(',').str[i]\n",
      "            del self.sample_df[col]\n",
      "            \n",
      "        self.sample_df_bytes = self.sample_df.values.nbytes + self.sample_df.index.nbytes + self.sample_df.columns.nbytes\n",
      "                \n",
      "        return 0\n",
      "        \n",
      "    #def get_sample_df(self):\n",
      "        \n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 33
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_df = vcf(vcf_path)\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 39
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_df.get_vcf_samples_df({'AD':2})\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 41,
       "text": [
        "0"
       ]
      }
     ],
     "prompt_number": 41
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_df.df.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th>CHROM</th>\n",
        "      <th>POS</th>\n",
        "      <th>ID</th>\n",
        "      <th>REF</th>\n",
        "      <th>ALT</th>\n",
        "      <th>QUAL</th>\n",
        "      <th>FILTER</th>\n",
        "      <th>INFO</th>\n",
        "      <th>FORMAT</th>\n",
        "      <th>NA12877</th>\n",
        "      <th>...</th>\n",
        "      <th>NA12880</th>\n",
        "      <th>NA12881</th>\n",
        "      <th>NA12882</th>\n",
        "      <th>NA12883</th>\n",
        "      <th>NA12884</th>\n",
        "      <th>NA12885</th>\n",
        "      <th>NA12886</th>\n",
        "      <th>NA12887</th>\n",
        "      <th>NA12888</th>\n",
        "      <th>NA12893</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CHROM</th>\n",
        "      <th>POS</th>\n",
        "      <th>REF</th>\n",
        "      <th>ALT</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th rowspan=\"5\" valign=\"top\">chrM</th>\n",
        "      <th>72 </th>\n",
        "      <th>T</th>\n",
        "      <th>G</th>\n",
        "      <td> chrM</td>\n",
        "      <td>  72</td>\n",
        "      <td> .</td>\n",
        "      <td> T</td>\n",
        "      <td> G</td>\n",
        "      <td> 9131.99</td>\n",
        "      <td> TruthSensitivityTranche99.90to100.00</td>\n",
        "      <td>                                   AC=2;AN=2;SF=2f</td>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>...</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>73 </th>\n",
        "      <th>G</th>\n",
        "      <th>A</th>\n",
        "      <td> chrM</td>\n",
        "      <td>  73</td>\n",
        "      <td> .</td>\n",
        "      <td> G</td>\n",
        "      <td> A</td>\n",
        "      <td> 8679.39</td>\n",
        "      <td> TruthSensitivityTranche99.90to100.00</td>\n",
        "      <td> AC=34;AF=1.00;AN=34;DP=250;DS;Dels=0.00;FS=0.0...</td>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td> 1/1:52:99:250:1.000:8638,737,0:99:0,249</td>\n",
        "      <td>...</td>\n",
        "      <td> 1/1:52:99:250:1.000:8365,731,0:99:0,250</td>\n",
        "      <td> 1/1:50:99:250:1.000:8644,740,0:99:0,250</td>\n",
        "      <td> 1/1:51:99:250:1.000:8448,737,0:99:0,250</td>\n",
        "      <td> 1/1:51:99:250:1.000:8752,737,0:99:0,250</td>\n",
        "      <td> 1/1:50:99:250:1.000:8788,743,0:99:0,250</td>\n",
        "      <td> 1/1:49:99:250:1.000:8757,740,0:99:0,250</td>\n",
        "      <td> 1/1:51:99:250:1.000:8433,731,0:99:0,250</td>\n",
        "      <td> 1/1:50:99:250:1.000:8628,728,0:99:0,250</td>\n",
        "      <td> 1/1:48:99:250:1.000:8812,746,0:99:0,250</td>\n",
        "      <td> 1/1:49:99:250:1.000:8699,743,0:99:0,250</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>146</th>\n",
        "      <th>T</th>\n",
        "      <th>C</th>\n",
        "      <td> chrM</td>\n",
        "      <td> 146</td>\n",
        "      <td> .</td>\n",
        "      <td> T</td>\n",
        "      <td> C</td>\n",
        "      <td> 9220.79</td>\n",
        "      <td> TruthSensitivityTranche99.90to100.00</td>\n",
        "      <td>                                   AC=2;AN=2;SF=2f</td>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>...</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "      <td>                                       .</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>150</th>\n",
        "      <th>T</th>\n",
        "      <th>C</th>\n",
        "      <td> chrM</td>\n",
        "      <td> 150</td>\n",
        "      <td> .</td>\n",
        "      <td> T</td>\n",
        "      <td> C</td>\n",
        "      <td> 9237.84</td>\n",
        "      <td> TruthSensitivityTranche99.90to100.00</td>\n",
        "      <td> AC=34;AF=1.00;AN=34;DP=250;DS;Dels=0.00;FS=0.0...</td>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td> 1/1:48:99:250:1.000:9231,746,0:99:0,250</td>\n",
        "      <td>...</td>\n",
        "      <td> 1/1:48:99:250:0.996:9001,717,0:99:1,249</td>\n",
        "      <td> 1/1:49:99:250:1.000:9111,746,0:99:0,249</td>\n",
        "      <td> 1/1:51:99:250:1.000:9227,752,0:99:0,250</td>\n",
        "      <td> 1/1:50:99:250:1.000:9344,749,0:99:0,250</td>\n",
        "      <td> 1/1:49:99:250:1.000:9322,752,0:99:0,250</td>\n",
        "      <td> 1/1:49:99:250:1.000:9265,749,0:99:0,250</td>\n",
        "      <td> 1/1:50:99:250:0.996:9180,714,0:99:1,249</td>\n",
        "      <td> 1/1:49:99:250:1.000:9319,752,0:99:0,250</td>\n",
        "      <td> 1/1:48:99:250:1.000:9204,749,0:99:0,250</td>\n",
        "      <td> 1/1:49:99:250:1.000:9217,746,0:99:0,250</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>152</th>\n",
        "      <th>T</th>\n",
        "      <th>C</th>\n",
        "      <td> chrM</td>\n",
        "      <td> 152</td>\n",
        "      <td> .</td>\n",
        "      <td> T</td>\n",
        "      <td> C</td>\n",
        "      <td> 9401.35</td>\n",
        "      <td> TruthSensitivityTranche99.90to100.00</td>\n",
        "      <td> AC=30;AF=1.00;AN=30;DP=250;DS;Dels=0.00;FS=0.0...</td>\n",
        "      <td> GT:MQ:GQ:DP:VF:PL:GQX:AD</td>\n",
        "      <td> 1/1:49:99:250:1.000:9429,746,0:99:0,250</td>\n",
        "      <td>...</td>\n",
        "      <td> 1/1:50:99:250:1.000:9381,746,0:99:0,250</td>\n",
        "      <td> 1/1:51:99:250:1.000:9343,746,0:99:0,249</td>\n",
        "      <td> 1/1:51:99:250:1.000:9331,746,0:99:0,250</td>\n",
        "      <td> 1/1:51:99:249:1.000:9392,740,0:99:0,248</td>\n",
        "      <td> 1/1:49:99:250:1.000:9421,746,0:99:0,250</td>\n",
        "      <td> 1/1:51:99:250:1.000:9409,746,0:99:0,250</td>\n",
        "      <td> 1/1:50:99:250:1.000:9399,743,0:99:0,250</td>\n",
        "      <td> 1/1:49:99:250:1.000:9432,749,0:99:0,249</td>\n",
        "      <td> 1/1:50:99:250:1.000:9316,746,0:99:0,250</td>\n",
        "      <td> 1/1:51:99:250:1.000:9372,746,0:99:0,250</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 26 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 51,
       "text": [
        "                  CHROM  POS ID REF ALT     QUAL  \\\n",
        "CHROM POS REF ALT                                  \n",
        "chrM  72  T   G    chrM   72  .   T   G  9131.99   \n",
        "      73  G   A    chrM   73  .   G   A  8679.39   \n",
        "      146 T   C    chrM  146  .   T   C  9220.79   \n",
        "      150 T   C    chrM  150  .   T   C  9237.84   \n",
        "      152 T   C    chrM  152  .   T   C  9401.35   \n",
        "\n",
        "                                                 FILTER  \\\n",
        "CHROM POS REF ALT                                         \n",
        "chrM  72  T   G    TruthSensitivityTranche99.90to100.00   \n",
        "      73  G   A    TruthSensitivityTranche99.90to100.00   \n",
        "      146 T   C    TruthSensitivityTranche99.90to100.00   \n",
        "      150 T   C    TruthSensitivityTranche99.90to100.00   \n",
        "      152 T   C    TruthSensitivityTranche99.90to100.00   \n",
        "\n",
        "                                                                INFO  \\\n",
        "CHROM POS REF ALT                                                      \n",
        "chrM  72  T   G                                      AC=2;AN=2;SF=2f   \n",
        "      73  G   A    AC=34;AF=1.00;AN=34;DP=250;DS;Dels=0.00;FS=0.0...   \n",
        "      146 T   C                                      AC=2;AN=2;SF=2f   \n",
        "      150 T   C    AC=34;AF=1.00;AN=34;DP=250;DS;Dels=0.00;FS=0.0...   \n",
        "      152 T   C    AC=30;AF=1.00;AN=30;DP=250;DS;Dels=0.00;FS=0.0...   \n",
        "\n",
        "                                     FORMAT  \\\n",
        "CHROM POS REF ALT                             \n",
        "chrM  72  T   G    GT:MQ:GQ:DP:VF:PL:GQX:AD   \n",
        "      73  G   A    GT:MQ:GQ:DP:VF:PL:GQX:AD   \n",
        "      146 T   C    GT:MQ:GQ:DP:VF:PL:GQX:AD   \n",
        "      150 T   C    GT:MQ:GQ:DP:VF:PL:GQX:AD   \n",
        "      152 T   C    GT:MQ:GQ:DP:VF:PL:GQX:AD   \n",
        "\n",
        "                                                   NA12877  \\\n",
        "CHROM POS REF ALT                                            \n",
        "chrM  72  T   G                                          .   \n",
        "      73  G   A    1/1:52:99:250:1.000:8638,737,0:99:0,249   \n",
        "      146 T   C                                          .   \n",
        "      150 T   C    1/1:48:99:250:1.000:9231,746,0:99:0,250   \n",
        "      152 T   C    1/1:49:99:250:1.000:9429,746,0:99:0,250   \n",
        "\n",
        "                                    ...                     \\\n",
        "CHROM POS REF ALT                   ...                      \n",
        "chrM  72  T   G                     ...                      \n",
        "      73  G   A                     ...                      \n",
        "      146 T   C                     ...                      \n",
        "      150 T   C                     ...                      \n",
        "      152 T   C                     ...                      \n",
        "\n",
        "                                                   NA12880  \\\n",
        "CHROM POS REF ALT                                            \n",
        "chrM  72  T   G                                          .   \n",
        "      73  G   A    1/1:52:99:250:1.000:8365,731,0:99:0,250   \n",
        "      146 T   C                                          .   \n",
        "      150 T   C    1/1:48:99:250:0.996:9001,717,0:99:1,249   \n",
        "      152 T   C    1/1:50:99:250:1.000:9381,746,0:99:0,250   \n",
        "\n",
        "                                                   NA12881  \\\n",
        "CHROM POS REF ALT                                            \n",
        "chrM  72  T   G                                          .   \n",
        "      73  G   A    1/1:50:99:250:1.000:8644,740,0:99:0,250   \n",
        "      146 T   C                                          .   \n",
        "      150 T   C    1/1:49:99:250:1.000:9111,746,0:99:0,249   \n",
        "      152 T   C    1/1:51:99:250:1.000:9343,746,0:99:0,249   \n",
        "\n",
        "                                                   NA12882  \\\n",
        "CHROM POS REF ALT                                            \n",
        "chrM  72  T   G                                          .   \n",
        "      73  G   A    1/1:51:99:250:1.000:8448,737,0:99:0,250   \n",
        "      146 T   C                                          .   \n",
        "      150 T   C    1/1:51:99:250:1.000:9227,752,0:99:0,250   \n",
        "      152 T   C    1/1:51:99:250:1.000:9331,746,0:99:0,250   \n",
        "\n",
        "                                                   NA12883  \\\n",
        "CHROM POS REF ALT                                            \n",
        "chrM  72  T   G                                          .   \n",
        "      73  G   A    1/1:51:99:250:1.000:8752,737,0:99:0,250   \n",
        "      146 T   C                                          .   \n",
        "      150 T   C    1/1:50:99:250:1.000:9344,749,0:99:0,250   \n",
        "      152 T   C    1/1:51:99:249:1.000:9392,740,0:99:0,248   \n",
        "\n",
        "                                                   NA12884  \\\n",
        "CHROM POS REF ALT                                            \n",
        "chrM  72  T   G                                          .   \n",
        "      73  G   A    1/1:50:99:250:1.000:8788,743,0:99:0,250   \n",
        "      146 T   C                                          .   \n",
        "      150 T   C    1/1:49:99:250:1.000:9322,752,0:99:0,250   \n",
        "      152 T   C    1/1:49:99:250:1.000:9421,746,0:99:0,250   \n",
        "\n",
        "                                                   NA12885  \\\n",
        "CHROM POS REF ALT                                            \n",
        "chrM  72  T   G                                          .   \n",
        "      73  G   A    1/1:49:99:250:1.000:8757,740,0:99:0,250   \n",
        "      146 T   C                                          .   \n",
        "      150 T   C    1/1:49:99:250:1.000:9265,749,0:99:0,250   \n",
        "      152 T   C    1/1:51:99:250:1.000:9409,746,0:99:0,250   \n",
        "\n",
        "                                                   NA12886  \\\n",
        "CHROM POS REF ALT                                            \n",
        "chrM  72  T   G                                          .   \n",
        "      73  G   A    1/1:51:99:250:1.000:8433,731,0:99:0,250   \n",
        "      146 T   C                                          .   \n",
        "      150 T   C    1/1:50:99:250:0.996:9180,714,0:99:1,249   \n",
        "      152 T   C    1/1:50:99:250:1.000:9399,743,0:99:0,250   \n",
        "\n",
        "                                                   NA12887  \\\n",
        "CHROM POS REF ALT                                            \n",
        "chrM  72  T   G                                          .   \n",
        "      73  G   A    1/1:50:99:250:1.000:8628,728,0:99:0,250   \n",
        "      146 T   C                                          .   \n",
        "      150 T   C    1/1:49:99:250:1.000:9319,752,0:99:0,250   \n",
        "      152 T   C    1/1:49:99:250:1.000:9432,749,0:99:0,249   \n",
        "\n",
        "                                                   NA12888  \\\n",
        "CHROM POS REF ALT                                            \n",
        "chrM  72  T   G                                          .   \n",
        "      73  G   A    1/1:48:99:250:1.000:8812,746,0:99:0,250   \n",
        "      146 T   C                                          .   \n",
        "      150 T   C    1/1:48:99:250:1.000:9204,749,0:99:0,250   \n",
        "      152 T   C    1/1:50:99:250:1.000:9316,746,0:99:0,250   \n",
        "\n",
        "                                                   NA12893  \n",
        "CHROM POS REF ALT                                           \n",
        "chrM  72  T   G                                          .  \n",
        "      73  G   A    1/1:49:99:250:1.000:8699,743,0:99:0,250  \n",
        "      146 T   C                                          .  \n",
        "      150 T   C    1/1:49:99:250:1.000:9217,746,0:99:0,250  \n",
        "      152 T   C    1/1:51:99:250:1.000:9372,746,0:99:0,250  \n",
        "\n",
        "[5 rows x 26 columns]"
       ]
      }
     ],
     "prompt_number": 51
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "Vcf_df = Vcf(vcf_path, 'NA12880')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "AttributeError",
       "evalue": "'Vcf' object has no attribute '_load_vcf'",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
        "\u001b[0;32m<ipython-input-50-8e9e7a420b89>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mVcf_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mVcf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvcf_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'NA12880'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[0;32m<ipython-input-47-9f570cf21eeb>\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, filename, sample_col, generate_features, autosome_only, no_halfs)\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0;31m# Load VCF\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_variant_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_load_vcf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m         \u001b[0;31m# Sample Column Name\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;31mAttributeError\u001b[0m: 'Vcf' object has no attribute '_load_vcf'"
       ]
      }
     ],
     "prompt_number": 50
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_df.sample_df.index.get_level_values(3).value_counts()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 30,
       "text": [
        "C      19605\n",
        "G      18796\n",
        "A      16561\n",
        "T      15765\n",
        ".       1182\n",
        "C,A      161\n",
        "AT       145\n",
        "CT       130\n",
        "CA       120\n",
        "C,G      117\n",
        "G,C      112\n",
        "T,G      110\n",
        "A,C      109\n",
        "AG       106\n",
        "TA       101\n",
        "...\n",
        "TTG                         1\n",
        "TCC                         1\n",
        "TCGGGGCGTGGCC               1\n",
        "TTTTTC                      1\n",
        "TAAAAAA,TAAAAA              1\n",
        "AAG                         1\n",
        "T,TT                        1\n",
        "AAGAC                       1\n",
        "CTT,C                       1\n",
        "AGAATG                      1\n",
        "TTGTAGTCTGACCTGTGGTCTGAC    1\n",
        "GCA                         1\n",
        "TGCCCA                      1\n",
        "CTTTGAT                     1\n",
        "AGG                         1\n",
        "Length: 244, dtype: int64"
       ]
      }
     ],
     "prompt_number": 30
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pd.set_option('io.hdf.default_format','table')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 128
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "h5 = '/Users/erickscott/git/pdVCF/pandasVCF/test_data/NA128_17member_pedigree_100k.h5'\n",
      "store = pd.HDFStore(h5, 'w', complevel=6, complib='blosc')\n",
      "#"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 275
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "store.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 274
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "store.create_table_index()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "hdf5_df = vcf_df.sample_df.reset_index().tail(5000)\n",
      "hdf5_df.set_index(['CHROM','POS'], inplace=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 243
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "store.append('sample_df',vcf_df.sample_df.reset_index(), data_columns=True, expectedrows=len(vcf_df.sample_df))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 276
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ids = vcf_df.sample_df.reset_index().tail(10000)[['POS']].astype(int)\n",
      "ids = list(set([i[0] for i in ids.values]))  #int values same as hdf5 store"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 277
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit pd.read_hdf(h5, 'sample_df', where='POS=ids and ALT=A')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 1.22 s per loop\n"
       ]
      }
     ],
     "prompt_number": 278
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Multi-Variant Query"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Benchmarking hdf5 reading using a list of ids\n",
      "pd.read_hdf(h5, 'sample_df', where='POS=ids and AD_1>125 and ALT=A')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 212,
       "text": [
        "count      8\n",
        "unique     3\n",
        "top       99\n",
        "freq       6\n",
        "dtype: object"
       ]
      }
     ],
     "prompt_number": 212
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "#STRUCTURE OF GENOME WIDE QUERY:\n",
      "1) STORE CHROMOSOMES IN DIFFERENT HDF5 DIRECTORIES (multi-index CHROM/data)\n",
      "2) INDEX ON POSITION\n",
      "3) DISTRIBUTE QUERIES BY CHROMOSOME\n",
      "4) RETURN ALL VARIANTS INTERSECTING POSITIONS AND ADDITIONAL CRITERIA (IF JOINING TO A LIST OF VARIANTS DO THIS IN MEMORY)\n",
      "5) RETURN AS AGGREGATED DF"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def var_query(h5_path, variant_id_df, where_criteria=''):\n",
      "    '''\n",
      "    This function accepts variant queries with a minimum of CHROM & POS\n",
      "    '''\n",
      "    \n",
      "    results = []\n",
      "    for i,chrom_group in variant_id_df.groupby(['CHROM']):\n",
      "        pos = chrom_group['POS'].unique().values\n",
      "        results.append(pd.read_hdf(h5_path, i, where='POS=pos and'+ where_criteria) )\n",
      "    return pd.concat(results)\n",
      "    \n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 214
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "store.select('sample_df',where='SAMPLE_ID=\"NA12878\" and AD_0>20 and AD_1>50 and ALT!=\".\"').info()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "<class 'pandas.core.frame.DataFrame'>\n",
        "Int64Index: 250 entries, 752728 to 802604\n",
        "Data columns (total 16 columns):\n",
        "CHROM        250 non-null object\n",
        "POS          250 non-null int64\n",
        "REF          250 non-null object\n",
        "ALT          250 non-null object\n",
        "SAMPLE_ID    250 non-null object\n",
        "DP           250 non-null object\n",
        "GQ           248 non-null object\n",
        "GQX          248 non-null object\n",
        "GT           248 non-null object\n",
        "MQ           250 non-null object\n",
        "VF           248 non-null object\n",
        "AD_0         250 non-null object\n",
        "AD_1         250 non-null object\n",
        "PL_0         248 non-null object\n",
        "PL_1         248 non-null object\n",
        "PL_2         248 non-null object\n",
        "dtypes: int64(1), object(15)"
       ]
      }
     ],
     "prompt_number": 45
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "a = pd.read_hdf(h5,'sample_df',where='AD_0>100')[['CHROM', 'POS', 'REF', 'ALT', 'AD_0']]\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 155
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "store.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 119
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!gzip /Users/erickscott/git/pdVCF/test_data/NA128_17member_pedigree_100k.h5\n",
      "!ls -lht /Users/erickscott/git/pdVCF/test_data/"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "total 61072\r\n",
        "-rw-r--r--  1 erickscott  staff    16M Aug 27 22:31 NA128_17member_pedigree_100k.h5\r\n",
        "-rw-r--r--  1 erickscott  staff    11K Aug 27 21:22 NA128_17member_pedigree_100k.vcf.gz.tbi\r\n",
        "-rw-r--r--  1 erickscott  staff    14M Aug 27 21:22 NA128_17member_pedigree_100k.vcf.gz\r\n"
       ]
      }
     ],
     "prompt_number": 104
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "store.root.sample_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 73,
       "text": [
        "/sample_df (Group) ''\n",
        "  children := ['table' (Table)]"
       ]
      }
     ],
     "prompt_number": 73
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_df.FORMAT.value_counts()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 8,
       "text": [
        "GT:GQ:DP:VF:PL:GQX:MQ:AD    94070\n",
        "GT:DP:MQ:AD                  2902\n",
        "GT:MQ:GQ:DP:VF:PL:GQX:AD     2438\n",
        "GT:MQ:DP:AD                    20\n",
        "GT:GQ:DP:VF:GQX:MQ:AD          10\n",
        "dtype: int64"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Groupby the FORMAT so that the splits will work\n",
      "genotypes = vcf_df.df[vcf_df.samples].groupby(by=vcf_df.FORMAT)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#For Clarity\n",
      "\n",
      "def vcf_samples_df(vcf_df, split_cols):\n",
      "    '''\n",
      "    This function creates a samples_df containing sample\n",
      "    level information for all non-missing variant calls\n",
      "    indexed on CHROM, POS, REF, ALT\n",
      "    \n",
      "    vcf_df\n",
      "    \n",
      "    split_cols\n",
      "    \n",
      "    '''\n",
      "        \n",
      "    genotypes = vcf_df.df[vcf_df.samples].groupby(by=vcf_df.FORMAT)  #genotypes grouped by FORMAT variant annotations\n",
      "    \n",
      "    #Iterate through genotype groups, dropping missing calls\n",
      "    master_df = []\n",
      "    for name,group in genotypes:\n",
      "        temp_group = genotypes.get_group(name)  #group of interest\n",
      "        del temp_group['FORMAT']  #remove the format column\n",
      "        temp_group.replace(to_replace='.', value=nan, inplace=True)  #replace . with none, allows stack to remove null columns, space savings\n",
      "        \n",
      "        temp_group = temp_group.stack()  #satcking samples for each variant\n",
      "    \n",
      "        #creating sample dataframe\n",
      "        temp_group_data = pd.DataFrame(temp_group.str.split(':').tolist())\n",
      "        temp_group_data.index = temp_group.index\n",
      "        temp_group_data.columns = name.split(':')\n",
      "        temp_group_data.replace(to_replace='.', value=nan, inplace=True)\n",
      "        \n",
      "        master_df.append(temp_group_data)\n",
      "    \n",
      "    #Concatenating all genotype groups\n",
      "    vcf_df.sample_df = pd.concat(master_df)\n",
      "    \n",
      "    \n",
      "    \n",
      "    #spliting user-defined columns\n",
      "    for col in split_cols:\n",
      "        for i in range(0, split_cols[col]):\n",
      "            vcf_df.sample_df[col + '_' + str(i)] = vcf_df.sample_df[col].str.split(',').str[i]\n",
      "        del vcf_df.sample_df[col]\n",
      "            \n",
      "    return 0"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "split_cols_dict = {'AD':2, 'PL':3}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 22
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(vcf_df.sample_df)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 31,
       "text": [
        "802611"
       ]
      }
     ],
     "prompt_number": 31
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_df.sample_df.values.nbytes + vcf_df.sample_df.index.nbytes + vcf_df.sample_df.columns.nbytes"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 27,
       "text": [
        "51367168"
       ]
      }
     ],
     "prompt_number": 27
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit vcf_samples_df(vcf_df, split_cols_dict)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 5.47 s per loop\n"
       ]
      }
     ],
     "prompt_number": 23
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "master_df_concat.reset_index().boxplot(['AD1','AD2'], notch=True, by='CHROM')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 33,
       "text": [
        "array([<matplotlib.axes.AxesSubplot object at 0x148a0ff10>,\n",
        "       <matplotlib.axes.AxesSubplot object at 0x14d1fa790>], dtype=object)"
       ]
      },
      {
       "metadata": {},
       "output_type": "display_data",
       "png": "iVBORw0KGgoAAAANSUhEUgAAAX8AAAEaCAYAAAD5fVeOAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XtYlGX6B/DvYKKmJrYVWIZ4IE5yGBUkUxndUtQ8V6yl\nidCaWOEhbX9blthWrm0n9bpEqrXU0tTKdDPFNF4Lc4EStDznMmpqaSWK5gl5fn9MvDLC4DAw88zj\n+/1c11zOe5iZe565vXnnfg9jEkIIEBGRofjIDoCIiDyPxZ+IyIBY/ImIDIjFn4jIgFj8iYgMiMWf\niMiAWPypRg0aNIDZbEZ0dDQGDBiA77//vl6fX9M0DBw4sMZ1Nm3ahC1bttTr63pCUFAQfvvttyrz\nmzVrVqfnzcrKQmRkJEJDQxETE4PFixcDACwWC7799lt9PavVisjISAC2cW7RogXMZjO6deuGefPm\n2T3nxo0bERsbi7Zt2yI2NhZffPGFviw5ORlNmzbF6dOn9XkTJ06Ej49Pte+P1MDiTzW6/vrrUVhY\niG3btiE5ORn/+Mc/PB5DTk4Ovv76a5cfL4SAjNNZTCZTreY7Y/Xq1Vi0aBHefPNN7N69G5qm4fff\nf9eft6bn7tmzJwoLC/HZZ59h1qxZOHHiBADg559/RmpqKp577jkUFxfj2WefRUpKCn755Rf9scHB\nwVi1ahUAoLy8HF988QVat27t8vsg+Vj8ySlCCPzyyy9o3LixPi8rKwvdunXDgAEDoGkaANsWYcUf\niOzsbCQkJEAIgeTkZEyePBmdO3eGxWJBXl5eldc4efIknn32WcTExCA1NRX79++H1WpFVlYWXn/9\ndZjNZuTm5to95sSJE5g8eTJCQ0ORnp6ub21brVaEhYVh7NixiIqKwqFDh6qN98pvHo8//jgWLlwI\nwLblPmPGDISFhSE5ORmHDx/WX3PGjBm46667cP/996OoqAgAUFJSgieffFKPpaY/OM888wxCQ0Mx\nceJElJSUYP/+/ejcubO+fN++fXbTFVasWIFnnnkGd955JwDAz88Pjz76qN3ndDV+fn6Ijo7G5s2b\nAQCrVq3CiBEj9HEYNGgQkpKS8MknnwCw/VFJSkrCsmXL9DHr3r07GjRocNXXIu/F4k81Onv2LMxm\nM9q2bYuMjAy89NJLAIDt27dj4cKFWLt2LV577TX89a9/BQDMnDkTy5YtQ05ODiZMmIB3331X3xr9\n5ptvoGkaXn75ZX39yhYuXIiTJ09i69at6N27N55++mkEBQVh3LhxmDx5MgoLC9G9e3e7x7z99tvw\n8fHBrl27YDabcfDgQX3Znj17MGDAAHz33XcoKSmpNt4rVd56NplMOHnyJHbu3ImIiAj885//BADM\nnj0bMTEx2Lx5M6ZPn45nnnkGALBgwQJcvHgRO3fuRHR0tF0slZ05cwY33XQTduzYgfLycvz73/9G\n+/bt0aJFC2zbtg0A8M477yAlJcXucRcuXMDnn3+OXr16Vfu8Qgg89NBDMJvNMJvNGDBgQLXfBKxW\nK/Lz8/WW0CeffII+ffrYrdO3b198/PHH+vQdd9yB48ePo6SkBB988AH+8pe/VBsDqYPFn2rUpEkT\nFBYWwmq1Yt68ebjvvvsAAJ9++inuu+8+tGjRAiEhIbjjjjuQl5eHJk2a4K233sI999yDJ554Am3b\ntgVgK6RDhw5F8+bNERcXByGEviVdYc2aNUhOToaPjw+SkpKwZcsWXLx4sca2zfr16/HQQw/BZDJh\nxIgRaNSokb7sT3/6EwYPHlxjvFczatQomEwmJCcnY/369QCAjz/+GBkZGTCbzRg1ahR27tyJs2fP\nYt26dXr8Dz30kF0slZlMJowePRoNGjTAww8/jHXr1gEAHnnkEbzzzjsoLy/H8uXL8eCDD141viuf\nd8mSJSgsLNTbO5XH7auvvkJ0dDQ6dOiACRMmoE2bNgCq/7YghKjyh2PYsGFYunQp8vLy0KNHj1rF\nRt6HxZ+cNmzYMOzatQtnzpyByWSqUjQqisX27dtx8803VynujtavaR1H6zkjICDA7jmqe/0mTZrg\n/Pnz+rxff/31qvFcunQJq1at0otscXExmjRp4nB9Zw0bNgxr167Fp59+ii5duqBly5Z2y319fdGn\nTx+7nbFXqvz6V8bSo0cPbNu2Ddu3b8e8efNw9OhRAMDQoUP1P2wV1q9fj6FDh+rTFa2f5557Dn36\n9KnTfgvyDiz+5LTNmzcjODgYTZs2xb333ouVK1fi5MmT2Lt3L/bt24e4uDgcOHAAr732GgoLC7F2\n7Vrk5+cDsBWiVatW4fTp0ygoKICPjw9uvfVWu+e/9957sXjxYly6dAkrVqxAt27d0LBhQ7Rp0wbH\njx+vNqa+fftiyZIlKC8vx7Jly+wK+ZXPXV280dHR2LlzJ06fPo3Dhw9XKYLvv/8+Ll26hEWLFqFv\n374AgAcffBBz587VX6ui59+vXz8sXrwY5eXlWLp0qcNYhBD6+1y8eDH69esHAGjcuDH69u2LtLQ0\njBkzptrH3nfffZg1a5b+reXUqVN466239OXOFOXw8HA8+eSTmD59OgBbj/+DDz7AmjVrIITAp59+\niuXLl+vfmipiDgwMxIsvvojx48df9TXI+7H4U40qev7R0dF4+eWX8dprrwEAIiMj8fDDD6Nfv36Y\nOHGiXoAeeeQRvPrqqwgICMC///1vPPLIIzh//jxMJhM6d+6MhIQETJ06VV+/co999OjRaN68OTp3\n7owNGzbo+xf69OmDb775BmazWd9JWSE1NRUXL15EeHg48vPz0a5dO32LuXIhdBRv48aN8dRTTyE+\nPh4pKSlVet833HADOnbsiO+++w7/93//B8C2U7hFixbo3r07IiIi8OabbwIAUlJS0KBBA4SHh2Pr\n1q16W+VKTZs2xbFjxxAREQGTyYTU1FR92YMPPggfH58qcVQYNGgQHnzwQaSmpiIsLAwWiwVNmzZ1\n+PlVHoPK98eNG4d169bhxx9/REBAAN566y1kZGSgbdu2mDFjBt5++23cfPPNVR47duxYu1YeqcvE\nSzqTJ4wZMwYDBw7EsGHD6vV5L1y4gAYNGqBBgwb46KOP8OGHH2Lp0qX18txt27bFt99+ixtvvLFe\nns8ZL730EoQQ+k5kInfhlr+XsVgsuPHGG3HhwgV9XnJyMho1aoQbbrgBAQEB6Nq1K15++WWcO3dO\nX+enn37CoEGDcNttt8HHx8fhkSbXmoMHDyI2NhYRERFYvXo1nnvuuXp7bk9v2Q4dOhRr167F5MmT\nPfq6srma82vWrEH37t3RsmVLdOnSBTNnzrRbTlchyGsUFxeLJk2aiJCQELFixQp9fnJysnj22WeF\nEEJYrVbx2muvCbPZLOLi4sSlS5eEEEL8/PPPIjMzU2zZskWYTCZx4MABKe+BqDbqkvNLliwR2dnZ\n4uzZs2Lv3r2iW7duIjMzU8r7UBG3/L3IokWLcPfdd2PUqFH6iUYVxB/duTZt2mDSpElYsWIFCgoK\n8M477wAAbrnlFowbNw5dunTxeNxErqpLzo8YMQJ9+vRB48aNERwcjPHjx+Pdd9/19FtQFou/F1m0\naBGSkpLwwAMPIDs72+ERLgDQvn17REVFoaCgwIMREtWv+sz5LVu2IDg42F2hXnNY/L1Ebm4uDh8+\njEGDBiE4OBjh4eF4//33a3xM9+7dcejQIQ9FSFS/6jPn161bh/fffx8vvviiu8K95rD4e4mFCxei\nT58+aN68OQDg/vvv178GCwcHZH355ZcIDAz0WIxE9am+cn7Lli0YOXIkVq5cyf8PtXCd7ADIdiz9\n8uXLUV5ejlatWgEAzp8/j5MnT2L79u3VXq1x//79+P777zFx4kQZIRPVSX3lfGFhIYYMGYKFCxfC\nYrF48i0oj8XfC3zyySe47rrrsG3bNvj6+gKwbfk88MADWLRokT4NAAcOHMAnn3yCxYsXo0uXLnZn\ngp47dw5lZWX6/XPnztldhZPIW9RHzn///fdITEzE3LlzMWDAADlvRGXyDjSiComJiWLKlClV5i9f\nvlwEBASIkSNHCl9fX9G8eXNxyy23iLi4ODFz5kxx9uxZu/VNJpMwmUzCx8dH/5fIG9VHzo8ZM0Y0\naNBANGvWTL917NjRk29DaU6d4Xvu3DkkJCTg/PnzaNy4MZKSkjBp0iSUlpZi5MiRKCwsRKdOnfDe\ne+/pv1I0Z84czJ07Fw0bNsSbb75Z5VK8REQkj9OXd/j9999x/fXX4/z58+jcuTNWrlyJlStX4tCh\nQ3jllVfw5JNPIigoCFOmTMGxY8fQs2dPrF+/HsXFxZg0aRK2bt3q7vdCREROcvpon+uvvx4AcPr0\naZSVlaFRo0bIz89HamoqGjVqhJSUFP1Kg3l5eUhMTERgYKD+S06lpaXueQdERFRrThf/8vJyREdH\nw9/fH48//jgCAwNRUFCA0NBQAEBoaKh++d68vDyEhYXpjw0JCdGXERGRfE4f7ePj44Nt27bBarWi\nf//+uOuuu2r1wxW8/CsRkfeo9aGeQUFB6N+/P/Ly8hAbG6v/duquXbsQGxsLAOjatSs2bNigP2b3\n7t36sgodOnTA/v376xg+Ue1ER0frP77iacx58rSa8t2pts8vv/yCkpISALafuVu/fj0GDx6Mrl27\nYsGCBTh79iwWLFiA+Ph4AEBcXByys7Nx8OBBaJoGHx8f/Sy+Cvv379d/m9Ubb9OnT5ceg4o3bx+3\nih9Il8Gbc97bPzdvvnnz2NWU705t+R89ehSjR4/GpUuXEBAQgClTpqBVq1ZIS0vDyJEjERISgk6d\nOmHWrFkAAH9/f6SlpaF3797w9fVFVlZW/fzvISKieuFU8Y+MjKz2UM3mzZtj1apV1T5mwoQJmDBh\nQt2ik8hqtcoOQUkcNzXxc3OdqmPHC7s5EBMTIzsEJXHc1MTPzXWqjp203/A1mUyQ9NJkYDLzjjlP\nnlZTznHLn4jIgFj8HdA0TXYISuK4qYmfm+tUHTsWfyIiA2LPnwyFPX8yEvb8iYjIDou/A6r28WTj\nuKmJn5vrVB07Fn8iIgNiz58MhT1/MhL2/ImIyA6LvwOq9vFk47ipiZ+b61QdOxZ/IiIDYs+fDIU9\nfzIS9vyJiMgOi78DqvbxZOO4qYmfm+tUHTsWfyIiA2LPnwyFPX8yEvb8iYjIDou/A6r28WTjuKmJ\nn5vrVB07Fn8iIgNiz58MhT1/MpI69/wPHTqEXr16ISIiAhaLBUuWLAEAZGRkoHXr1jCbzTCbzVi7\ndq3+mDlz5iA4OBjh4eHIzc2th7dBRCTHtGnAyy87t+7vvwMxMUBxsXtjqjPhhKNHj4rCwkIhhBDH\njx8Xbdu2FadOnRIZGRni1VdfrbL+zz//LEJCQsSBAweEpmnCbDZXWcfJl5YmJydHdghK8vZxk5l3\n3pzz3v65yTZ2rBCNGgmxb1/VZVeO3dNPCwEIUVTkmdhqUlPOObXlHxAQgJiYGADATTfdhIiICBQU\nFFT88aiyfl5eHhITExEYGIiEhAQIIVBaWlpPf648o6hIdgREntOrl+wIvF+7dsBjjwE1de527gTe\nfBO49VbPxeWqWu/w/eGHH7Bjxw507doVADB37lzEx8dj1qxZeoHPz89HWFiY/piQkBDk5+fXU8ie\nUVJikR2CkiwWi+wQyCUW2QF4vfHjgSNHgBUr7OdX5LwQQFoaMH06cNNNno+vtmpV/EtLS5GUlITX\nX38dTZs2RVpaGoqLi5GdnY39+/cjKysLQPXfBkwmU/1E7CH//a/sCIjIW/j6Av/4B/D990BSEtCq\nFRAaan/z8QG+/BKYPRvYvdv2GG92nbMrXrx4EcOHD8eoUaMwePBgAMAtt9wCAGjRogUee+wxjB8/\nHlOmTEHXrl2xYcMG/bG7d+9GbGxsledMTk5GUFAQAMDPzw8xMTH6X9GKY2c9OV1UdHmLPzv7DSQn\nxyAoyALbKp6PR8XpinneFI+mabBarfAG3pTzJpP2R1QWABoub59ZIIT35JQ3TM+cCcTGarhwAfjr\nXy04+pMJ2k/QWVBRIYCA/wg0bAgcPKjh5589XcOKUFJSAgBXz3lndhqUl5eLUaNGiUmTJtnNP3Lk\niBBCiIsXL4qnnnpKvPDCC0IIIX766Sd9h29OTo6SO3yBHNkhKMnbdxzKzDtvy3l/f9uOSdstR7/v\n7y87Mu81fboQQ4faz6vI+f37hfjTn4QoLvZ0VI7VlHNOHeefm5uLnj17IioqSm/fvPTSS1i6dCmK\niorg6+uLnj17Ytq0abjxxhsBALNnz8bcuXPh6+uLrKws9OjRw+45vfGY5zfeAD75xHZ/0yYgIcF2\nf8gQYOJEeXFR/eFx/pfV1In1ojC9xp49wF132Q4Gad26+nVeegn4+mvgP/+peXw9paac40lelURG\nArt22e5fugQ0aGC7HxYGfPedvLio/rD4X+bjU32RN5mA8nLPx+PNhADuvhu4915g0iTH6124YDvG\n/8UXgaFDPRefI7ywm5PmzrWdzDFtGgBo+v25c2VHpo7KvXbybn5+lac0B/MJAD74APj1V+CJJ6ou\nq5zzvr5AZiaQng6cOeO5+Fzh9A5fI3jiictb/gDwwgu2fz/6iFv+dO05caJ2841M04CxY4HrnKiY\nCQlAixbADz8A0dFuD81lLP6VtGhx+cO9dMmi32/RQl5Mqqk48oBUY5EdgNeraANfqbqcd7SuN2Hx\nr+S++y4X/02bgPh42/0hQ+TFROQuDRrY9m1VN5+ufez5V/L3v9uK/qZNAKDp9//+d9mRqYM9f3XY\n79TVHMynq1E151n8K0lMtLV4Kto8FfcTE+XGReQOjRrVbj5dW9j2qaTiGH8bC06erG4+1YQ9f3Wc\nO1d5yuJgPl2NqjnPLX8iIgNi8XdIkx2AklTtfxqR/RmomoP5dDWq5jyLP5FBOTrZ2ItOQiY3Ys/f\nIYvsAJRR0+W6velyBlQTi+wAlMWePxmWEEK/AeKKafJWLVvWbj5dW1j8HdJkB6AoTXYA5CT7yzho\nDubT1bDnT0RKCQqy7dyt6NpV3P/jt2boGseev0MW2QEoyiI7AHKS/Q89WfQdvV7yo2fKYM+fiIiU\nweLvkCY7ACWNHq3JDoFcoskOQFns+RMBSE6WHQEROYPF3yGL7ACUpGr/kyyyA1CWqjnP4k9EZEAs\n/g5psgNQkqr9T9JkB6AsVXOexZ+IyICcKv6HDh1Cr169EBERAYvFgiVLlgAASktLMXjwYAQGBmLI\nkCE4ffq0/pg5c+YgODgY4eHhyM3NdU/0bmWRHYCSNM0iOwRyiUV2AMq6pnv+DRs2xOuvv44dO3bg\nww8/xLRp01BaWorMzEwEBgZi3759aN26NebPnw8AOHbsGObNm4eNGzciMzMT6enpbn0T5D1mzJAd\nARE5w6niHxAQgJiYGADATTfdhIiICBQUFCA/Px+pqalo1KgRUlJSkJeXBwDIy8tDYmIiAgMDkZCQ\nACEESktL3fcu3EKTHYCiNNkBkEs02QEoyzA9/x9++AE7duxAXFwcCgoKEBoaCgAIDQ1Ffn4+AFvx\nDwsL0x8TEhKiLyMiIvlqVfxLS0uRlJSE119/Hc2aNavVJXtruua7d7LIDkBRFtkBkEsssgNQlqo9\nf6cv7Hbx4kUMHz4co0aNwuDBgwEAsbGx2LVrF8xmM3bt2oXY2FgAQNeuXbFhwwb9sbt379aXVZac\nnIygPy4h6Ofnh5iYGH0gK75KeXr68n8C+2lZ8ag27W3jVXHf6iVXK/OmnL8yx5nzNU/v2aNB05xf\nv6BAw4kTno23qKgIJSUlAHD1nBdOKC8vF6NGjRKTJk2ymz9r1izx+OOPi99//12MHz9e/Otf/xJC\nCPHTTz+JkJAQceDAAZGTkyPMZnOV53TypT3K9gN2Fbccu2lyzujRObJDqJHMvPO2nGe+O2/sWCHm\nz69+WU5OTpV5UVFCFBW5NyZn1JRzTm35b968Ge+99x6ioqJgNpsBADNnzkRaWhpGjhyJkJAQdOrU\nCbNmzQIA+Pv7Iy0tDb1794avry+ysrKceRm6BvDaPkRqMP3x18HzL2wyed3P/NW0W8LLQiUXycw7\nb8t55rvzHn0U6NTJ9q8zoqOBRYts/8pUU87xDF8iIgNi8XdIkx2AkirvaCWVaLIDUJaqOc/iT0Rk\nQCz+DllkB6AkXttHVRbZASjr8qGzamHxp3rFa/sQqYHF3yFNdgCK0mQHQC7RZAegLPb8iYhIGSz+\nDllkB6Aoi+wAyCUW2QEoiz1/IiJSBou/Q5rsAJQ0erQmOwRyiSY7AGWx508EXtuHSBUs/g5ZZAeg\nJFX7n2SRHYCyVM15Fn8iIgNi8XdIkx2AklTtf5ImOwBlqZrzLP5ERAbE4u+QRXYASuK1fVRlkR2A\nstjzJwKv7UOkChZ/hzTZAShKkx0AuUSTHYCy2PMnIiJlsPg7ZJEdgKIssgMgl1hkB6AsVXv+18kO\ngNR0+jRw8WL1y06cqDqvWTOgYUP3xkREzmPxd0gDt4aq9+uvQECAraBfqWFDDe3aWezmXbgADB8O\nLFrkkfDIJRqY767RNE3JrX+n2j4pKSnw9/dHZGSkPi8jIwOtW7eG2WyG2WzG2rVr9WVz5sxBcHAw\nwsPDkZubW/9Rk1TnzgG33GLbwr/ytn591XmLFgFnzsiOmogqc6r4jxkzBuvWrbObZzKZMHnyZBQW\nFqKwsBD9+vUDABw7dgzz5s3Dxo0bkZmZifT09PqP2iMssgNQkopbQAQw312nas471fbp0aMHrFZr\nlflCiCrz8vLykJiYiMDAQAQGBkIIgdLSUjRv3rzOwRIRUf2o09E+c+fORXx8PGbNmoXS0lIAQH5+\nPsLCwvR1QkJCkJ+fX7copdBkB6AkVY95Jk12AMpSNedd3uGblpaG5557DqdOncLUqVORlZWFKVOm\nVPttwGQyVfscycnJCAoKAgD4+fkhJiZG/wpVMaCenr789bfIblpWPN46ff68Bk2rbvxQ7frHj1e/\nvvs/T9v96r65yuBNOX+54Fc/LTvHvG16z57a5XxBgYYTJzwbb1FREUpKSgDg6jkvnFRcXCw6duxY\n7bKioiLRrVs3IYQQq1evFunp6fqy6OhocerUqSqPqcVLewzg+EaX/fijELfe6vz6H34oxLBh7oun\nNmTmnbflPPPdeWPHCjF/vvPrR0UJUVTkvnicVVPOudz2OXr0KACgrKwMS5YsQf/+/QEAcXFxyM7O\nxsGDB6FpGnx8fNjvJyLyMk4V/xEjRqBbt27Ys2cPbr/9dixYsAB/+9vfEBUVhfj4eFy8eBFpaWkA\nAH9/f6SlpaF3794YP348Zs+e7dY34D6a7ACUdOVXYVKFJjsAZama8071/JcuXVplXkpKisP1J0yY\ngAkTJrgeFRERuRWv7eOQRXYASrq8M5HUYpEdgLJUzXkWfyIiA2Lxd0iTHYCSVO1/kiY7AGWpmvMs\n/kREBsTi75BFdgBKUrX/SRbZAShL1Zxn8SciMiAWf4c02QEoSdX+J2myA1CWqjnP4k9EZEAs/g5Z\nZAegJFX7n2SRHYCyVM15Fn8iIgNi8XdIkx2AklTtf5ImOwBlqZrzLP5ERAbE4u+QRXYASlK1/0kW\n2QEoS9WcZ/EnIjIgFn+HNNkBKEnV/idpsgNQlqo5z+JPRGRALP4OWWQHoCRV+59kkR2AslTNeRZ/\nIiIDYvF3SJMdgJJU7X+SJjsAZama8yz+REQGxOLvkEV2AEpStf9JFtkBKEvVnGfxJyIyIKeKf0pK\nCvz9/REZGanPKy0txeDBgxEYGIghQ4bg9OnT+rI5c+YgODgY4eHhyM3Nrf+oPUKTHYCSVO1/kiY7\nAGWpmvNOFf8xY8Zg3bp1dvMyMzMRGBiIffv2oXXr1pg/fz4A4NixY5g3bx42btyIzMxMpKen13/U\nRERUJ04V/x49eqBly5Z28/Lz85GamopGjRohJSUFeXl5AIC8vDwkJiYiMDAQCQkJEEKgtLS0/iN3\nO4vsAJSkav+TLLIDUJaqOe9yz7+goAChoaEAgNDQUOTn5wOwFf+wsDB9vZCQEH0ZERF5h+tcfaAQ\nwul1TSZTtfOTk5MRFBQEAPDz80NMTIz+V7Sij+bp6ctbQG8AiNGnZcXjrdPnz2vQtOrGzzbvyvWP\nH69+ffd/nrb7VqsV3sCbcv5yn99S6X7FtPwc87bpPXtql/MFBRpOnPBsvEVFRSgpKQGAq+e8cFJx\ncbHo2LGjPj1s2DCxdetWIYQQ33zzjRg+fLgQQojVq1eL9PR0fb3o6Ghx6tSpKs9Xi5f2GKDyLcdu\nmi778Uchbr21+mU5OTlV5n34oRDDhrk3JmfJzDtvy3nmu/PGjhVi/vzql1WX81FRQhQVuTcmZ9SU\ncy63fbp27YoFCxbg7NmzWLBgAeLj4wEAcXFxyM7OxsGDB6FpGnx8fNC8eXNXX0Yii+wAlHR5q5LU\nYpEdgLJUzXmniv+IESPQrVs37N27F7fffjveeecdpKWl4eDBgwgJCcHhw4cxbtw4AIC/vz/S0tLQ\nu3dvjB8/HrNnz3brGyAiotpzque/dOnSauevWrWq2vkTJkzAhAkTXI/KK2jg1lDtaZqm7JaQsWlg\nvrtG1Zx3eYcvEZFRlJQAR47Yblf65Zeq8y9e9ExcdcHi75BFdgBKUnELiADme82WL7f9+9ZbVZeV\nllpw5W7Nhg2BG290f1x1weJPRHQVNR3ZbjIBKp7Hygu7OaTJDkBJlY99JpVosgNQmCY7AJew+BMR\nGRCLv0MW2QEoiT1/VVlkB6Awi+wAXMLiT0RkQCz+DmmyA1ASe/6q0mQHoKzRozXZIbiExZ+IqA6S\nk2VH4BoWf4cssgNQEnv+qrLIDkBZquY8iz8RkQGx+DukyQ5ASez5q0qTHYCyVM15Fn8iIgNi8bdj\nqnRLuGKanKFq/9OYLgEQf9wsle5fkhiTejTNIjsEl7D4VyKE0G+A6YppomuLEA0ghAlCmP6YNv1x\nayA5MrXMmCE7Atew+DukyQ5ASar2P0mTHYDCNNkBuITF34GcHNkREBG5D4u/A6r28WRjz19VFtkB\nKMwiOwAe4QlaAAAPeUlEQVSXsPg7oGofj4jIGSz+DmmyA1ASe/6q0mQHoCxe24eIyIB4bZ9rjkV2\nAEpiz19VFtkBKEvVnK9z8Q8KCkJUVBTMZjPi4uIAAKWlpRg8eDACAwMxZMgQnD59us6BEpH7TJ8u\nOwLytDoXf5PJBE3TUFhYiPz8fABAZmYmAgMDsW/fPrRu3Rrz58+vc6CepmofTzb2/NVksWiyQ1CW\nqjlfL22fK8+Azc/PR2pqKho1aoSUlBTk5eXVx8t4lKp9PCIiZ9TLln/v3r0xZMgQrF69GgBQUFCA\n0NBQAEBoaKj+jUAlqvbxZOO4qYmfm+tUPSfouro+webNm9GqVSvs2rULAwcORFxcnNPXwklOTkZQ\nUBAAwM/PDzExMXoSVnyV4rR3Tp8/r0HTnF//+PHarV9f0xX3rVYrvAFz/tqbnjHDgowM74inqKgI\nJSUlAHDVnDeJerxq2eTJkxEWFoZ169Zh2rRpMJvN+PbbbzFz5kx8+OGH9i9sMnn1BdM0TdMHlewd\nPgzExdn+vVJ14/bRR8CSJbZ/ZZOZd96c88x315lMGoSwyA6jWjXlXJ3aPr///jtKS0sBAMePH0d2\ndjYSExPRtWtXLFiwAGfPnsWCBQsQHx9fl5chIjd7913ZEZCn1an4//zzz+jRowdiYmLwl7/8BU8+\n+SRuv/12pKWl4eDBgwgJCcHhw4cxbty4+orXY1Tt48nGrUc1LVxokR2CwiyyA3BJvbZ9avXCXvwV\nGABMJsCLw5OqprZPddj2kf/aV8N8d503j53b2j7XNk12AEqqvKOVVKLJDkBZqp4TxOJPRFQHqp4T\nVOdDPa9dFtkBKIk9f+9VVmZrv124UN1SC95+u+rcuDggKsrdkalN1Zw3dPE/dQo4csTx8t277acb\nNQLatnVvTETuUlwMPP448MADVZeZzcB//2s/b98+YNMmYPFiz8RHnmXo4v/EE8DnnwM33FB12Q03\naBgyxGI3b/9+wGoFbrvNI+EpiceLezd/f1S7hV/d57Z4MbB+vWfiUpmqOW/o4n/hAvDqq8CIEVWX\n2c5GtZ/Xtq2jr8xERGrhDl8HVPxL7g04bmri5+Y6Vc8JYvEnIqoDVX/vm8XfAR6v7hqOm5r4udWF\nJjsAl7D4ExEZkKGL/+7dwNixwJ//DDz4IHDmzOVlFT3QZctsy//8Z9uRPuXlUkJVBnvHauLnVhcW\n2QG4xNBH++zcaTt654svbNNLlpqqrJP0x80E2/UxeLQPqerMGeCHH2zXogGAPn2A7Gz7dX74AQgO\ntp/H4/yvTYbe8h82zHbGoxB/XJip4o4Q0HJy7KaFAIKCgMaNZUft3dg79l7XX2/7d8YMICMDyF5v\nsv0lMJmg/fFvh2ATBEzIyAC6d7f9gaCaqXptH0Nv+RMZickEdOgAPPfcHzOmV7ra4xUntkwH0K4d\nT/JyhqrX9jH0ln9N2AN1DcdNTfzcXKfq2LH4ExEZENs+Dqh6vQ7ZOG7e69w52w7drVurLvvmGw1d\nuljs5nnJb957PVVznsWfyCBOnrT9+8gjVZdZrbYDGq40apQ7IyKZWPwdUPEvuTfguHmv7t0d/9yg\nyWTBb795Np5rhaZZqlwEUgWGLv4ffGC7LVjg3PpHjwI+3EtCRJVUHDqrGkMX//fes52xGxBQddnL\nL2t46imL3bxnnwUCAz0Tm7crK7NtRZquOC+uuv5nWZnn4iJXaVD1TFUZTFckfuVJRz+Y7m3cVvy/\n/PJLPProoygrK0N6ejqeeOIJd72Uyx56yPGyPn1sP/RCVfn7227LlwNJSTWvW14OzJkDjB7tmdiI\nPKFygVd1h69JuOnPlNlsxuzZs9GmTRv07dsXubm5uOmmmy6/sMnk1X8hTSbH/VECtmwBhg+3XSLD\nz8/xellZwMKFQG6ud7TMZOadN+c88/3aVFPOueW/48k/Divo2bMn2rRpgz59+iAvL88dL0WS3Hkn\nMHgw8Pe/O17np5+AadNsfwC8ofCTY9Ony46APM0t/yULCgoQGhqqT4eHh+O/V/46tNfTZAfg9WbO\nBFavBr7++vK8ytf2mTjRdlhhZKTnY6PasVg02SEoS9XrWRl6hy+55sAB2zHhEycCx44Bd91luw8A\nhw4Bq1YBe/YAa9faLpk9aRLQqxcwaJDUsImoErcU/9jYWEydOlWf3rFjBxITE6usl5ycjKA/zizx\n8/NDTEyMvuOk4q+pJ6d79eplF1/lPfg5OTkej8dbp8+dAwAN584BZWUWCJigvQHbcts/0AA8BWBy\ngUBUFPDKKxpuuMHz8Vbct3rJ6arelPNXHrFSmRDCq3KO085NFxUVoaSkBACumvNu3+EbGBiIxMRE\n5Xb4kmOHDwNxcbZ+/4oVtgtCVldHCgqAgQOB55+3XTf+o488HmoV3OFLRuLxHb4A8MYbb+DRRx/F\n3XffjfHjx9sVfhVU3nKkqk6dsp3cMn++feGvPG6xscADD9iKP3k35rvrVB07t/X8ExISsGvXLnc9\nPUl2+rStlx8WVvN6L7wAfPyxZ2IiIue5re1z1RfmV2BllZYCI0faft/YmV82++wz25Ukp01zf2xX\nw7YPGUlNOcfiT4bC4k9GIqXnrzpV+3iycdzUxM/NdaqOHYs/EZEBse1DhsK2DxkJ2z5ERGSHxd8B\nVft4snHc1MTPzXWqjh2LPxGRAbHnT4bCnj8ZCXv+RERkh8XfAVX7eLJx3NTEz811qo4diz8RkQGx\n50+Gwp4/GQl7/kREZIfF3wFV+3iycdzUxM/NdaqOHYs/EZEBsedPhsKePxkJe/5ERGSHxd8BVft4\nsnHc1MTPzXWqjh2LPxGRAbHnT4bCnj8ZCXv+RERkx+Xin5GRgdatW8NsNsNsNmPt2rX6sjlz5iA4\nOBjh4eHIzc2tl0A9TdU+nmwcNzXxc3OdqmPncvE3mUyYPHkyCgsLUVhYiH79+gEAjh07hnnz5mHj\nxo3IzMxEenp6vQXrSUVFRbJDUBLHTU383Fyn6thdV5cHV9dLysvLQ2JiIgIDAxEYGAghBEpLS9G8\nefO6vJTHlZSUyA5BSRw3NfFzc52qY1ennv/cuXMRHx+PWbNmobS0FACQn5+PsLAwfZ2QkBDk5+fX\nLUoiIqpXNRb/e+65B5GRkVVuq1evRlpaGoqLi5GdnY39+/cjKysLQPXfBkwmk3uidyOr1So7BCVx\n3NTEz811yo6dqAdFRUWiW7duQgghVq9eLdLT0/Vl0dHR4tSpU1UeEx0dLQDwxptHb9HR0fWR8i5h\nzvPm6VtN+e5yz//o0aNo1aoVysrKsGTJEvTv3x8AEBcXh6lTp+LgwYP43//+Bx8fn2r7/aruJCFy\nFXOevInLxf9vf/sbioqK4Ovri549eyItLQ0A4O/vj7S0NPTu3Ru+vr56O4iIiLyHtDN8iYhIHsOf\n4ZucnIyPPvrIqXVTUlLg7++PyMhIN0fl/Wozbj4+Phg1apQ+XVZWhptvvhkDBw50V3hUA+Z87V2L\n+W744u/MkUhlZWUAgDFjxmDdunXuDkkJtRm3pk2bYseOHTh37hwA4PPPP0fr1q2VPArsWsCcr71r\nMd8NV/yzs7MxePBgxMTE4OGHHwYAFBQUoFu3bujSpQs2bNgAwHbKdq9evTB8+HB9q6dHjx5o2bKl\ntNhlcmXcoqKi9Mf3798fa9asAQAsXboUI0aM4EXOPIQ5X3uGyHcPHeXmFc6cOSPat28v9u7dK4QQ\n4rfffhOjR48Wffv2FWfPnhW5ubmiV69eQgghcnJyhI+Pj9i6davdcxQXF4uOHTt6PHaZ6jpuzZo1\nE9u3bxf33XefOHfunIiJiRGapol7771XyvsxEuZ87Rkl3w215b9mzRrcfffdCA4OBgC0bNkSJpMJ\n999/Pxo3bow777wTW7du1dePiYmB2WyWFa7XqI9xi4yMhNVqxdKlSzFgwACPxm9kzPnaM0q+1+na\nPioS1Xz18vPzA2DbUXPp0iV9/q233uqxuLxdfYzboEGDMGXKFGzatAnHjx93T6BUBXO+9oyQ74ba\n8h8wYAA2bNiAvXv3AgB+++03yRGpob7GLSUlBRkZGYiIiKjP8KgGzPnaM0q+G2rL//rrr0dmZiYm\nTZqEH3/8EZ06dYLJZLLbC+/oPgCMGDECmzZtwq+//orbb78dzz//PMaMGeOx+GWp67hVTN922214\n/PHH9XnedvTDtYg5X3tGyXee5EVEZECGavsQEZENiz8RkQGx+BMRGRCLPxGRAbH4ExEZEIs/EZEB\nsfgTERkQi78bWa1WNGnSBJ06dQIAnDhxAmPGjEG7du0QERGBAQMGYN++fbBarVWul56RkYFXX30V\ngO1a4u3atUNMTAyGDRtmd12RsrIyPPPMM7jjjjvQoUMHjBgxAr/++qu+/GrXFl+2bBmCg4O97lrj\npCbmvDpY/N2sQ4cOeuKmpqYiICAAeXl52LFjB5599lkcOXKk2jP/Kp8RaDKZ8Morr6CoqAhJSUl4\n6aWX9PXeeOMN7Ny5E9nZ2fj222/Rvn17uzMwr3Zt8aSkJLz99ttue/9kPMx5NbD4e8jp06exdetW\nzJw5EzfffDMAID4+HgkJCdVeREoIYTe/4v7dd9+NnJwcff7KlSvxz3/+E23btkWLFi3wwgsvYPv2\n7Thz5oy+ztWuLc6TvMkdmPPejcXfQz777DP06NHD4fL9+/fDbDbrt6ysrCpbR0IILF++HDExMQCA\no0eP4siRIwgJCbFbr2fPnnriA7YtnQ8++ADnz5/Hd999h65du9bjOyOqHnPeuxnqwm4yXe2iTu3b\nt0dhYaE+PWPGDH3rRAiBqVOnYvLkyTh+/Di+//77Wr2eCtcWp2sPc967ccvfQ/r164evvvqqVo+5\nsv954MABTJs2DU899RQAoFWrVmjVqhX27NmjP0YIga+++gr9+/e3e66Ka4t75c/J0TWJOe/dWPw9\npFmzZujUqROmTZum/7BDQUEBvvzyS4ePqa5H+fTTT+N///sftmzZAgAYNmwYnn76aVitVpw8eRLT\np09Hx44d0bRpU7vn8vZri9O1hznv3Vj8Pejtt9/Gjz/+iLi4OHTs2BHPP/88brvtNgDVf0V2dM3w\nadOmYcaMGQCASZMmISQkBPfccw86deqEvXv34t13363yuJquLe5t1xmnawdz3nvxev5uZLVaMXDg\nQHz33XeyQ6mRpml49dVX8Z///Ed2KKQ45rw6uOXvRtdddx1Onjypn/DijZYtW4bHHnsMN954o+xQ\n6BrAnFcHt/yJiAyIW/5ERAbE4k9EZEAs/kREBsTiT0RkQCz+REQG9P9wGhMIraJdgQAAAABJRU5E\nrkJggg==\n",
       "text": [
        "<matplotlib.figure.Figure at 0x148a07c10>"
       ]
      }
     ],
     "prompt_number": 33
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df = pd.read_table('/Users/erickscott/Dropbox/raslseq/Bcell_exp2/Analysis/Heuristic/Lane1_2_aggregated_counts.txt',sep=\"\\t\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 71
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df_groups = df.groupby('PlateBarcode')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 73
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df_groups['>M13_2///>M13_2'].aggregate(np.sum)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 81,
       "text": [
        "PlateBarcode\n",
        "ACCTAGG         142262\n",
        "ACTTCAA         192664\n",
        "AGTTGGT         511761\n",
        "ATCTTGC          39944\n",
        "CAGGTCG          77415\n",
        "CATATTG          62506\n",
        "CATCGAG         455533\n",
        "CCAATGC          42492\n",
        "CCTGCGA         189850\n",
        "CGAGATC          43161\n",
        "CGCATTA          72479\n",
        "CGGAGTT          95578\n",
        "GACGTCA          10646\n",
        "GAGATTC         100887\n",
        "GAGCATG          67521\n",
        "GATCCAA          54487\n",
        "GGACGCA         116223\n",
        "GGTACCT          97058\n",
        "GTAATTG           2780\n",
        "GTACCGG          80467\n",
        "GTTGCGT          56615\n",
        "TCTCCAT           7394\n",
        "TGATAGT          41154\n",
        "TGCAGAG         255846\n",
        "TGGCATC           9550\n",
        "TTCGAGC         650513\n",
        "TTGATCC          30941\n",
        "Name: >M13_2///>M13_2, dtype: float64"
       ]
      }
     ],
     "prompt_number": 81
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}